// webhook/.cov/run/sqlscan/res/page/common/header/header.js
var $ = require('jquery');
var Container = require('../../../com/container/container');
var context = require('../context');
var Store = require('../../../core/store');
var eventProxy = require('../../../core/eventproxy');
var Helper = require('../helper/index');
var Header = function(){};

Container._extend(Header, {
  _className : 'header',
  __filename: __filename,
  isHide : false,
  isStretchHide : false,
  isHelperHide : false,
  fixedOnScroll : true,
  helpUrl : '',
  _loadStore : function(done){
    var userInfo, param,  _self = this;
    this.store = Store._create();
    userInfo = context.options.session.user;
    param = {
      "requestId": new Date().getTime().toString(),
      "loginId": userInfo.WorkId
    };
    param = JSON.stringify(param) + '/' + userInfo.WorkId;
    this.store.load({
      'url' : context.urls.getApplicationByUser,
      'param' : {
        "paramKey": param
      },
      'callback': function(err, data){
        _self.selfApplist = data;
        done();
      }
    });
  },
  afterComplete : function(){
    Header.prototype._super.afterComplete.call(this);
    var _self = this;
    var subMenu, selectSubMenu, hash;
    // this._loadUserAppList(this.selfApplist);
    if(!this.isHelperHide && !this.isStretchHide){
      new Helper();
    }
    if(!this.fixedOnScroll){
      var prefix = _self._getPrefix();
      this._getHeaderInfo(prefix, function(err, data){
        var hash = localStorage.getItem('dxp_toolbar_notices');
        var notices = data.notices;
        if(notices.hash !== hash){
          var btnMsg = _self.query("*[name=message]");
          btnMsg.el.find('.btn').addClass('notify');
        }
      });
    }
    subMenu = $('.sub-menu', this.el);
    if(subMenu.is(':visible')){
      var controller = context.options.controller;
      if(controller === 'cheetah/document'){
        controller = 'cheetah/document/guide';
      }
      selectSubMenu = subMenu.find('a[href="/s/' + controller + '"]');
      // if(selectSubMenu.length === 0){
      //   window.location.href = subMenu.find('a').eq(0).attr('href');
      // }
      this._selectSubMenu(selectSubMenu, selectSubMenu.parent().parent(), hash);
    }
  },
  /**_loadUserAppList : function(data){
    var item, arr, _item, _self = this;
    var appId = localStorage.getItem('appId');
    if(data.length > 0){
      var count = 0;
      for (var i = data.length - 1; i >= 0; i--) {
        item = data[i];
        if(item.type !== 0){
          count ++;
          arr = [];
          if(count <= 9){
            arr.push("<div class='app-item'>");
            arr.push("<img src='/img/app.png'>");
            arr.push("<div class='content'>");
            arr.push("<span class='title'>" + item.name + "</span>");
            arr.push("</div>");
            arr.push("</div>");
            _item = $(arr.join('')).bind('click', {appObj:item, sel: _self}, _self._selectApp);
            _self.appList.append(_item);
          }
        }
        if(appId && item.id === parseInt(appId)){
          $('.curapp>span').html("<label title='" + item.name +"'>"+ item.name +"</label><i></i>");
        }
      }
      if(count > 10){
        _self.appList.append("<a class='app-more' href='/s/cheetah/app'>更多</a>");
      }
      eventProxy.on('header.appchange', function(obj){
        $('.curapp>span').html("<label title='" + obj.appName +"'>"+ obj.appName +"</label><i></i>");
      });
    }else{
      $('.curapp>span').attr("data", "-1");
      $('.curapp>span').html("<label title='暂无应用'>暂无应用</label><i></i>");
    }
  },
  _selectApp: function(target){
    var data = target.data.appObj;
    var _self = target.data.sel;
    var appId = data.id;
    localStorage.setItem('appId', appId);
    $('.curapp>span').html("<label title='"+ data.name +"'>"+ data.name +"</label><i></i>");
    var sel = $('.nav-items .active', _self.el).attr('data');
    if(sel === 'cheetah/app'){
      eventProxy.emit('menu.switch', {
        controller : 'manage',
        app : appId
      });
      _self.appList.hide();
    }else{
      window.location.href = '/s/cheetah/app#controller:manage|app:' + appId;
    }
  },**/
  initData: function(){
    this.hash = this.router || {};
    var navConfig, searchConfig;
    navConfig = [
      {
        alias : '首页',
        value : 'cheetah',
        // href : 'http://dw-pre.alibaba-inc.com:8000/index.htm'
        href : '/s/cheetah'
      },
      {
        alias : '应用开发',
        value : 'cheetahApp',
        // href : 'http://dw-pre.alibaba-inc.com:8000/app/index.vm'
        href : '/s/cheetah/app'
      },
      {
        alias : '数据管理',
        value : 'metadata',
        href : '/s/metadata'
      },
      {
        alias : '社区',
        value : 'atatech',
        href : context.options.atatech
      },
      {
        alias : '文档',
        value : 'cheetahDocument',
        href : '/s/cheetah/document/guide'
      }
      ];
    searchConfig = ["数据市场", "应用开发"];
    this.data = {
      config: navConfig,
      searchConfig: searchConfig,
      userInfo: context.options.session.user,
      nick: context.options.session.user.nickNameCn || context.options.session.user.nick,
      helpUrl : this.helpUrl
    };
  },
  afterRender: function(){
    Header.prototype._super.afterRender.call(this);
    var _self = this;
    _self._nav = $('.container', _self.el);
    _self._searchInput = $('.search-input', _self.el);
    _self._searchItems = $('.bar-items', _self.el);
    _self._searchItemChidren = $('ul', _self._searchItems);
    _self._logoInfo = $('.logo-info', _self.el);
    _self._logoNewInfo = $('.logo-newinfo', _self.el);
    _self._globalNav = $('.global-nav', _self.el);
    _self.gbm = $('.gbm', _self.el);
    _self._minLogo = $('.min-logo', _self._globalNav);
    _self._navSearch = $('.nav-search', this._globalNav);
    _self._topInfo = $('.info', _self._globalNav);
    _self.placeholder = $('.placeholder_div', _self.el);
    _self.appList = $('.app-list', _self.el);
    _self.highLightIndex = -1;
    _self.timer = undefined;
    _self.gbmTimer = undefined;
    _self.appTimer = undefined;
    _self.offsetConstant = 10;
    _self._activeNav();
  },
  initEvents: function(){
    var _self = this;
    var target, _class;
    Header.prototype._super.initEvents.call(this);
    _self.el.on('click', function(e){
      target = e.target;
      _class = target.className;
      switch(_class){
        case 'userinfo':
          _self.showUserInfo();
          break;
        case 'loginout':
          break;
      }
    });
    _self._searchInput.on('keydown', function(e){
      _self._searchItems.show();
    });
    _self._searchInput.on('keyup', function(e){
      var evt, keycode, _thisValue;
      evt = e || window.event;
      keycode = evt.keyCode || e.which;
      _thisValue = $(this).val();
      _self._searchTextKeyUp(keycode, _thisValue);
    });
    _self._searchInput.on('click', function(){
      if($(this).val().length > 0){
        _self._searchItems.show();
      }
    });
    _self._searchItems.on('mouseleave', function(){
      var $this = $(this);
      _self.timer = setTimeout(function(){
        $this.hide();
      }, 500);
    });
    _self._searchItems.on('mouseenter', function(){
      clearTimeout(_self.timer);
    });
    _self._searchItemChidren.delegate("li", "mouseover", function(){
      _self._changeAllLiToWhite();
      $(this).css('background-color','#D6E9F8');
      _self.highLightIndex = $(this).attr("data-index");
    });
    _self._searchItemChidren.delegate("li", "click", function(){
      // 跳转
    });
    //设置消息退出
    _self.el.delegate('.curinfo>span', 'click', function(e){
      var target = $(e.target).parent();
      var offset = target.offset();
      _self.gbm.css({
        'top' : offset.top + target.height() + _self.offsetConstant,
        'left' : offset.left
      });
      _self.gbm.show();
    });
    _self.gbm.on('mouseenter', function(){
      if(_self.gbmTimer){
        clearTimeout(_self.gbmTimer);
      }
    });
    _self.gbm.on('mouseleave', function(){
      var _this = $(this);
      _self.gbmTimer = setTimeout(function(){
        _this.hide();
      }, 500);
    });
    //应用列表
    _self.el.delegate('.curapp>span', 'click', function(e){
      var target = $(e.target).parent();
      if($(target).attr('data') !== '-1'){
        var offset = target.offset();
        var width = _self._topInfo.width();
        var right = '0px';
        if (!_self.isStretchHide){
          right = (_self.getClientWidth() - _self._logoInfo.width()) / 2;
        }
        _self.appList.css({
          'top' : offset.top + target.height() + _self.offsetConstant,
          'right' : right
        });
        _self.appList.show();
      }
    });
    _self.appList.on('mouseenter', function(){
      if(_self.appTimer){
        clearTimeout(_self.appTimer);
      }
    });
    _self.appList.on('mouseleave', function(){
      var _this = $(this);
      _self.appTimer = setTimeout(function(){
        _this.hide();
      }, 500);
    });
    if(_self.isHide){
      _self._logoNewInfo.hide();
      _self._hideLogoInfo();
      _self.emit('hide');
    }else if(_self.isStretchHide){
      _self._hideLogoInfo();
      _self._logoNewInfo.hide();
      _self.emit('hide');
      _self._nav.css({width: '100%', margin: '0px 5px'});
    }else if(_self.fixedOnScroll){
      _self._logoNewInfo.hide();
      _self._scroll();
    }else{
      _self._globalNav.css({'position':'relative'});
      _self._logoInfo.hide();
      _self._logoNewInfo.show();
      _self.docEvent();
      _self.notice();
      _self.settingEvent();
    }
    $('.document-menu a', this.el).on('click', function(){
      _self._selectSubMenu($(this), $(this).parent().parent());
    });
    this.logo = $('.logo', this.el);
    this.logo.on('click', function(){
      window.location.href = _self._getPrefix() +'/s/cheetah';
    });
  },
  _selectSubMenu : function(el, parent, hash){
    parent.find('a').removeClass('sub-active');
    el.addClass('sub-active');
    // eventProxy.emit('menu.switch', hash || {
    //   controller : el.attr('data-controller')
    // });
  },
  _hideLogoInfo: function(){
    var _self = this;
    var infoHeight, originalTop, logoInfoUl, width;
    infoHeight = _self._topInfo.height();
    originalTop = (_self._globalNav.height() - infoHeight) / 2;
    logoInfoUl = $('ul', _self._logoInfo);
    $('.nav-items li:first-child').remove();
    width = logoInfoUl.width();
    _self._minLogo.show();
    _self._logoInfo.hide();
    _self._topInfo.show();
    _self._globalNav.addClass("navfixed");
    _self._navSearch.animate({'right': width}, 20);
  },
  _searchTextKeyUp: function(keycode, value){
    var _self = this;
    var curText;
    if(keycode === 38 || keycode === 40){
      _self._processKeyUpAndDown(keycode);
    }else if(keycode === 13){
      if(_self.highLightIndex != -1){
        var s = _self._searchItemChidren.children().eq(_self.highLightIndex);
        // TODO
      }
    }else{
      $("li", _self._searchItemChidren).each(function(index, item){
        $(this).attr("data-value", value);
        curText = "<font style='color:#888'> 在</font>" + _self.data.searchConfig[index] + "<font style='color:#888'>中搜索 ";
        $(this).html(curText + "<font class='keyword'> " + _self._subStr(value, 10) + " </font>");
      });
      _self.highLightIndex = 0;
      _self._searchItemChidren.children().eq(_self.highLightIndex).css('background-color','#D6E9F8');
    }
  },
  _processKeyUpAndDown: function(key){
    var words, num;
    words = this._searchItemChidren.children();
    num = words.length;
    if(num <= 0){
      return;
    }
    this._changeToWhite();
    this.highLightIndex = ((key != 38 ? num + 1:num - 1) + this.highLightIndex) % num;
    words.eq(this.highLightIndex).css('background-color','#D6E9F8');
  },
  _changeAllLiToWhite: function(){
    var _self = this;
    $('li', _self._searchItemChidren).each(function(index, item){
      $(this).css('background-color', 'white');
    });
  },
  _changeToWhite: function(){
    if(this.highLightIndex != -1){
      this._searchItemChidren.children().eq(this.highLightIndex).css('background-color','white');
    }
  },
  _activeNav: function(){
    var _this, _self = this;
    $('.nav-items li', this._nav).each(function(i, item){
      _this = $(this);
      if(context.options.app === _this.attr('data')){
        _this.addClass('active');
        _self.el.find('.global-nav').addClass(_this.attr('data'));
      }
    });
  },
  _scroll: function(){
    var _self = this;
    $(window).bind('scroll', function(){
      var scrollTop = _self.getScrollTop();
      _self.fnScroll(scrollTop);
    });
  },
  fnScroll: function(scrollTop){
    var _self = this;
    var logoInfoUl, width;
    var originalTop = $('.info', _self.el).offset().top;
    var originalHeight = _self._logoInfo.height();
    logoInfoUl = $('ul', _self._logoInfo);
    width = logoInfoUl.width();
    if(scrollTop <= 0){
      _self._globalNav.removeClass("navfixed");
      _self._navSearch.animate({'right': '0px'}, 20);
      _self._logoInfo.css({"visibility": "visible", "height": "64px"});
      _self._minLogo.hide();
      _self._topInfo.hide();
      $('.nav-items li:first-child').show();
      _self.placeholder.css({'height': '37px'});
    }else{
      _self._globalNav.addClass("navfixed");
      _self._navSearch.animate({'right': width}, 20);
      _self._logoInfo.css({"visibility": "hidden", "height": "37px"});
      _self.placeholder.css({'height': '0px'});
      $('.nav-items li:first-child').hide();
      _self._minLogo.show();
      _self._topInfo.show();
    }
  },
  showUserInfo : function(){
    this.gbm.hide();
    var win, _self = this;
    root.loadJs('/module/page/common/userinfo/index.js', function(err, objs){
      win = objs['/module/page/common/userinfo/index.js']._create({
        title: '个人信息',
        width: '600px',
        height:'450px'
      });
    });
  },
  _subStr: function(str, length){
    var len = 0, strLength = str.length;
    if(strLength > length){
      for (var i = 0; i < strLength; i++) {
        len += str.charCodeAt(i) > 255 ? 2 : 1;
        if (len > length) {
          return str.substr(0, i) + '...';
        }
      }
    }else{
      return str;
    }
  },
  getScrollTop: function() {
    var scrollTop = 0;
    if (document.documentElement && document.documentElement.scrollTop) {
      scrollTop = document.documentElement.scrollTop;
    }else if (document.body) {
      scrollTop = document.body.scrollTop;
    }
    return scrollTop;
  },
  getClientWidth: function(){
    if(window.innerWidth){
      winWidth = window.innerWidth;
    }else if((document.body) && (document.body.clientWidth)){
      winWidth = document.body.clientWidth;
    }
    return winWidth;
  },
  _getPrefix: function(){
    return context.options.prefix;
  },
  docEvent: function(){
    var _self = this;
    _self.btnDoc = this.query("*[name=doc]");
    var opt = null, offset = null, docNav = null, popoverBody = null;
    offset = this.btnDoc.el.offset();
    opt = {
      left: offset.left - 23,
      top: offset.top + _self.btnDoc.el.outerHeight(true)
    };
    var fn = function(obj){
      var propover = require('../../../newcom/popover/index.js')._create({
        events: {
          'aftercomplete': function(){
            popoverBody = $('.popover-content', propover.el);
            docNav = require('../docnav/index.js')._create({
              renderTo: popoverBody,
              prefix: obj.prefix,
              atatech : obj.atatech,
              events: {
                'aftercomplete' : function(){
                  propover.show(opt);
                  propover.emit('docnav.loaded');
                }
              }
            });
          }
        }
      });
      return propover;
    };
    this.btnDoc.on('click', function(){
      var btn = _self.btnDoc.el.find('.btn');
      var _prefix = _self._getPrefix();
      var atatech = context.options.atatech;
      if(btn.hasClass('selected')){
        _self.propoverDestroy();
      }else{
        _self.propoverDropdown();
        var obj = {
          prefix: _prefix,
          atatech : atatech
        };
        _self.propover = fn(obj);
        $(window).bind('click.dropup', function(e){
          if(btn.hasClass('selected') && !_self.el.find(e.target).length){
            _self.propoverDestroy();
            _self.btnDoc.el.find('.doc').trigger('click');
          }
        });
      }
    });
    this.btnDoc.el.find("button.js-dropdown").on('dropup', function(){
      _self.propoverDestroy();
      _self.btnDoc.el.find('.doc').removeClass('selected');
    });
  },
  propoverDropdown: function(){
    $('button.js-dropdown').trigger('dropup');
  },
  propoverDestroy: function(){
    var _self = this;
    if(_self.propover){
      _self.propover.destroy();
      _self.propover = null;
    }
    $(window).unbind('click.dropup');
  },
  notice: function(){
    var _self = this;
    _self.btnMsg = this.query("*[name=message]");
    var opt = null, offset = null,
    notify = null, popoverBody = null, flag = false,loginAll, left, right, bodyWidth, minLeft, popoverWidth, btnWidth;
    loginAll = $('.loginAll');
    if(loginAll.offset()){
      left = loginAll.offset().left + loginAll.outerWidth(true);
    }
    bodyWidth = $(document.body).outerWidth(true);
    right = bodyWidth - left;
    offset = this.btnMsg.el.offset();
    btnWidth = this.btnMsg.el.width();
    var fn = function(){
      var popoverWidth;
      var propover = require('../../../newcom/popover/index.js')._create({
        events: {
          'aftercomplete': function(){
            popoverBody = $('.popover-content', propover.el);
            notify = require('../notify/index.js')._create({
              renderTo: popoverBody,
              events: {
                'aftercomplete': function(){
                  var prefix = _self._getPrefix();
                  _self._getHeaderInfo(prefix, function(err, data){
                    if(!err){
                      eventProxy.emit('toolbar.notices', null, data.notices);
                    }
                  });
                  opt = {
                    top : offset.top + _self.btnMsg.el.outerHeight(true),
                    right: right
                  };
                  propover.show(opt);
                  popoverWidth = popoverBody.width();
                  minLeft = bodyWidth - popoverWidth - right;
                  propover.setArrowPosition(offset.left - minLeft + btnWidth/3);
                  propover.emit('notify.loaded');
                }
              }
            });
          }
        }
      });
      return propover;
    };
    this.btnMsg.on('click', function(){
      var btn = _self.btnMsg.el.find('.btn');
      btn.removeClass('notify');
      if(btn.hasClass('selected')){
        _self.propoverDestroy();
      }else{
        _self.propoverDropdown();
        _self.propover = fn();
        $(window).bind('click.dropup', function(e){
          if(btn.hasClass('selected') && !_self.el.find(e.target).length){
            _self.propoverDestroy();
            _self.btnMsg.el.find('.mail').trigger('click');
          }
        });
      }
    });
    this.btnMsg.el.find("button.js-dropdown").on('dropup', function(){
      _self.propoverDestroy();
      _self.btnMsg.el.find('.mail').removeClass('selected');
    });
  },
  settingEvent: function(){
    var _self = this;
    _self.btnStting = this.query("*[name=setting]");
    var opt = null, offset = null, settingdom = null,
    popoverBody = null, flag = false, loginAll, left, bodyWidth, right, minLeft;
    loginAll = $('.loginAll');
    if(loginAll.offset()){
      left = loginAll.offset().left + loginAll.outerWidth(true);
    }
    bodyWidth = $(document.body).outerWidth(true);
    right = bodyWidth - left;
    offset = this.btnStting.el.offset();
    var fn = function(prefix){
      var popoverWidth;
      var propover = require('../../../newcom/popover/index.js')._create({
        events: {
          'aftercomplete': function(){
            popoverBody = $('.popover-content', propover.el);
            settingdom = require('../settings/index.js')._create({
              renderTo: popoverBody,
              prefix: prefix,
              events: {
                'aftercomplete' : function(){
                  opt = {
                    top : offset.top + _self.btnStting.el.outerHeight(true),
                    right: right
                  };
                  var btnWidth = _self.btnStting.el.width();
                  propover.show(opt);
                  popoverWidth = popoverBody.width();
                  minLeft = bodyWidth - popoverWidth - right;
                  propover.setArrowPosition(offset.left - minLeft + btnWidth/2);
                  propover.emit('settingdom.loaded');
                }
              }
            });
          }
        }
      });
      return propover;
    };
    this.btnStting.on('click', function(){
      var btn = _self.btnStting.el.find('.btn');
      var _prefix = _self._getPrefix();
      if(btn.hasClass('selected')){
        _self.propoverDestroy();
      }else{
        _self.propoverDropdown();
        _self.propover = fn(_prefix);
        $(window).bind('click.dropup', function(e){
          if(btn.hasClass('selected') && !_self.el.find(e.target).length){
            _self.propoverDestroy();
            _self.btnStting.el.find('.setting').trigger('click');
          }
        });
      }
    });
    this.btnStting.el.find("button.js-dropdown").on('dropup', function(){
      _self.propoverDestroy();
      _self.btnStting.el.find('.setting').removeClass('selected');
    });
  },
  _getHeaderInfo: function(prefix, cb){
    var url = prefix + '/nr/header/view';
    $.ajax({
      type: 'GET',
      url: url,
      dataType:'jsonp',
      async: false,
      success: function(data){
        if(data){
          data = data.replace(/\n|\r/g, '');
          data = data.replace(/\\/g, '\\\\');
          var obj = JSON.parse(data);
          cb(null, obj);
        }
      },
      error: function(jqXHR, textStatus, errorThrown){
        cb(errorThrown);
      }
    });
  }
});

exports = module.exports = Header;
