// webhook/run/sqlscan/res/page/createresource/createresource.js
var WindowPanel = require('hc-common/panel/windowpanel');
var Container = require('hc-common/panel/container');
var Store = require('hc-common/core/store');
var context = require('../../temp/context');
var Msg = require('hc-common/msg/msg');
var CreateResource = function(){};
WindowPanel._extend(CreateResource, {
  _className : 'createresource',
  width : '550px',
  height : '400px',
  record : null,
  afterRender : function(){
    CreateResource.prototype._super.afterRender.call(this);
    this._container = WinContainer._create({
      renderTo : this.body,
      win : this
    });
  }
});
var WinContainer = function(){};
Container._extend(WinContainer, {
  _className : 'createsourcewin',
  record : null,
  _loadTpls : function(done){
    var _self = this;
    root.loadTpl('/page/createresource/', 'createresource', function(err, tpls){
      _self.tpl = tpls.createresource;
      done();
    });
  },
  _loadCss : function(done){
    this._cssprefix = '/page/createresource/';
    root.loadCss(this._cssprefix, ['createresource'], function(){
      done();
    });
  },
  initData : function(){
    /*this.data = this.attr;
    this.msg = this.errMsg;*/
  },
  initEvents : function(){
    var _self = this;
    var upload = $('.sqlScanUploadFile>input');
    var type = _self.query('combobox[name=scriptType]');
    upload.on('change', function(){
      if($(this).val() || _self.fileValueTest){
        if(type.val() && _self.checkFileType(upload, type.val())){
          return;
        }
        $('.sqlScanUploadFile>label').html('');
        //filesizeTest测试的时候传
        var size = this.files ? this.files[0].size : _self.fileValueTest;
        var limit = 10 * 1024 * 1024;
        if(size > limit){
          $(this).val('');
          $('.sqlScanUploadFile>label').html('* 文件不能超过10M');
        }
      }else{
        $('.sqlScanUploadFile>label').html('* 请选择文件');
      }
    });
    type.on('change', function(){
      if(upload.val() && _self.checkFileType(upload, this.val())){
        return false;
      }
    });
    var btnCreateResourceSure = this.query('button[name="btnCreateResourceSure"]');
    btnCreateResourceSure.on('click', function(){
      _self.createResourceAction();
    });
    var btnCreateResourceCancel = this.query('button[name="btnCreateResourceCancel"]');
    btnCreateResourceCancel.on('click', function(){
      _self.win.doClose();
    });
  },
  afterComplete : function(){
    WinContainer.prototype._super.afterComplete.call(this);
    this.loadScriptTypeData();
  },
  checkFileType : function(upload, scriptType){
    var fileType = upload[0].files ? upload[0].files[0].name.split('.') : this.fileValueTest;
    if(fileType.length === 1 || fileType[fileType.length-1] != scriptType){
      //$(upload[0]).val('');
      $('.sqlScanUploadFile>label').html('* 文件需与脚本类型相同');
      Msg.alert('文件类型需与脚本类型相同,当前应传入以".' + scriptType + '"结尾的文件名');
      return true;
    }
    return false;
  },
  loadScriptTypeData : function(){
    var _self = this;
    var type = _self.query('combobox[name=scriptType]');
    var data = [/*{
      value : 'jar',
      text : 'java'
    }, */{
      value : 'py',
      text : 'python'
    }, {
      value : 'pl',
      text : 'perl'
    }, {
      value : 'sh',
      text : 'shell'
    }];
    type.store.load(data);
  },
  createResourceAction : function(){
    var _self = this;
    var userInfo = context.options.session.user;
    var name = this.query('field[name=scriptName]');
    var file = $('.sqlScanUploadFile>input');
    var type = this.query('combobox[name=scriptType]');
    var desc = this.query('textareafield[name=scriptDesc]');
    if(!name.validate() || !file.val() || !type.validate() || !desc.validate()){
      this.checkValueTest = '参数输入不正确';
      return;
    }else if(this.checkFileType(file, type.val())){
      return;
    }
    this.win.loading(true);
    var reader = new window.FileReader();
    reader.readAsText(file[0].files[0]);
    reader.onload = function(e){
      var fileContent = _self.base64Encode(this.result);
      var opt = {
        type : 'POST',
        url : context.urls.uploadResourceTotfs,
        param : {
          fileContent : fileContent
        }
      };
      var params = {
        resourceInfos : JSON.stringify({
          name : $.trim(name.val()),
          type : 2,//资源类型，脚本or配置，目前只支持脚本
          detail : $.trim(desc.val()),
          author : userInfo.nickNameCn,
          authorUserName : userInfo.odpsDisplayName
        }),
        type : type.val()
      };
      _self.uploadResourceTotfs(opt, params);
    }
    if(this.checkValueTest){//checkValueTest为测试mock数据
      reader.result = 'test';
      reader.onload();
      this.checkValueTest = null;
    }
  },
  sendRequest : function(opt, cb){
    var store = Store._create(opt);
    store.load(function(err, data){
      cb(err, data);
    });
  },
  uploadResourceTotfs : function(opt, params){
    var _self = this;
    this.sendRequest(opt, function(err, data){
      if(err){
        _self.win.loading(false);
        return;
      }
      params.scriptName = data.name;
      var opts = {
        url : context.urls.insertResource,
        param : params
      };
      _self.submitResource(opts);
    });
  },
  submitResource : function(opt){
    var _self = this;
    this.sendRequest(opt, function(err, data){
      if(err){
        _self.win.loading(false);
        return;
      }
      _self.win.doClose();
      var resourceLabel = $('#sqlScanResourceId');
      resourceLabel.attr("resourceId", data.id);
      resourceLabel.attr("value", data.name);
      Msg.alert('资源创建成功');
      _self.checkValueTest = '资源创建成功';
      _self.emit('aftercreateresource');
    })
  },
  base64Encode : function(input){
    var _keyStr = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=";
    var output = "";
    var chr1, chr2, chr3, enc1, enc2, enc3, enc4;
    var i = 0;
    input = this._utf8_encode(input);
    while (i < input.length) {
      chr1 = input.charCodeAt(i++);
      chr2 = input.charCodeAt(i++);
      chr3 = input.charCodeAt(i++);
      enc1 = chr1 >> 2;
      enc2 = ((chr1 & 3) << 4) | (chr2 >> 4);
      enc3 = ((chr2 & 15) << 2) | (chr3 >> 6);
      enc4 = chr3 & 63;
      if (isNaN(chr2)) {
        enc3 = enc4 = 64;
      }else if (isNaN(chr3)) {
        enc4 = 64;
      }
      output = output + _keyStr.charAt(enc1) + _keyStr.charAt(enc2) +
      _keyStr.charAt(enc3) + _keyStr.charAt(enc4);
    }
    return output;
  },
  _utf8_encode : function (string) {
    //string = string.replace(/\r\n/g,"\n");
    var utftext = "";
    for (var n = 0; n < string.length; n++) {
      var c = string.charCodeAt(n);
      if (c < 128) {
        utftext += String.fromCharCode(c);
      }else if((c > 127) && (c < 2048)) {
        utftext += String.fromCharCode((c >> 6) | 192);
        utftext += String.fromCharCode((c & 63) | 128);
      }else {
        utftext += String.fromCharCode((c >> 12) | 224);
        utftext += String.fromCharCode(((c >> 6) & 63) | 128);
        utftext += String.fromCharCode((c & 63) | 128);
      }
    }
    return utftext;
  }/*,
  // public method for decoding  
  base64Decode : function (input) { 
    var _keyStr = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/="; 
    var output = "";  
    var chr1, chr2, chr3;  
    var enc1, enc2, enc3, enc4;  
    var i = 0;  
    input = input.replace(/[^A-Za-z0-9\+\/\=]/g, "");  
    while (i < input.length) {  
      enc1 = _keyStr.indexOf(input.charAt(i++));  
      enc2 = _keyStr.indexOf(input.charAt(i++));  
      enc3 = _keyStr.indexOf(input.charAt(i++));  
      enc4 = _keyStr.indexOf(input.charAt(i++));  
      chr1 = (enc1 << 2) | (enc2 >> 4);  
      chr2 = ((enc2 & 15) << 4) | (enc3 >> 2);  
      chr3 = ((enc3 & 3) << 6) | enc4;  
      output = output + String.fromCharCode(chr1);  
      if (enc3 != 64) {  
          output = output + String.fromCharCode(chr2);  
      }  
      if (enc4 != 64) {  
          output = output + String.fromCharCode(chr3);  
      }  
    }  
    output = this._utf8_decode(output);  
    return output;  
  },
   _utf8_decode : function (utftext) {  
    var string = "";  
    var i = 0;  
    var c = c1 = c2 = 0;  
    while ( i < utftext.length ) {  
      c = utftext.charCodeAt(i);  
      if (c < 128) {  
          string += String.fromCharCode(c);  
          i++;  
      } else if((c > 191) && (c < 224)) {  
          c2 = utftext.charCodeAt(i+1);  
          string += String.fromCharCode(((c & 31) << 6) | (c2 & 63));  
          i += 2;  
      } else {  
          c2 = utftext.charCodeAt(i+1);  
          c3 = utftext.charCodeAt(i+2);  
          string += String.fromCharCode(((c & 15) << 12) | ((c2 & 63) << 6) | (c3 & 63));  
          i += 3;  
      }  
    }  
    return string;  
  }*/    
});
exports = module.exports = CreateResource;
