// webhook/.cov/lib/project.js
/* ****** automatically generated by jscover - do not edit ******/
if (typeof _$jscoverage === 'undefined') { _$jscoverage = {}; }
/* ****** end - do not edit ******/
function BranchData() {
    this.position = -1;
    this.nodeLength = -1;
    this.src = null;
    this.evalFalse = 0;
    this.evalTrue = 0;

    this.init = function(position, nodeLength, src) {
        this.position = position;
        this.nodeLength = nodeLength;
        this.src = src;
        return this;
    }

    this.ranCondition = function(result) {
        if (result)
            this.evalTrue++;
        else
            this.evalFalse++;
    };

    this.pathsCovered = function() {
        var paths = 0;
        if (this.evalTrue > 0)
          paths++;
        if (this.evalFalse > 0)
          paths++;
        return paths;
    };

    this.covered = function() {
        return this.evalTrue > 0 && this.evalFalse > 0;
    };

    this.toJSON = function() {
        return '{"position":' + this.position
            + ',"nodeLength":' + this.nodeLength
            + ',"src":' + jscoverage_quote(this.src)
            + ',"evalFalse":' + this.evalFalse
            + ',"evalTrue":' + this.evalTrue + '}';
    };

    this.message = function() {
        if (this.evalTrue === 0 && this.evalFalse === 0)
            return 'Condition never evaluated         :\t' + this.src;
        else if (this.evalTrue === 0)
            return 'Condition never evaluated to true :\t' + this.src;
        else if (this.evalFalse === 0)
            return 'Condition never evaluated to false:\t' + this.src;
        else
            return 'Condition covered';
    };
}

BranchData.fromJson = function(jsonString) {
    var json = eval('(' + jsonString + ')');
    var branchData = new BranchData();
    branchData.init(json.position, json.nodeLength, json.src);
    branchData.evalFalse = json.evalFalse;
    branchData.evalTrue = json.evalTrue;
    return branchData;
};

BranchData.fromJsonObject = function(json) {
    var branchData = new BranchData();
    branchData.init(json.position, json.nodeLength, json.src);
    branchData.evalFalse = json.evalFalse;
    branchData.evalTrue = json.evalTrue;
    return branchData;
};

function buildBranchMessage(conditions) {
    var message = 'The following was not covered:';
    for (var i = 0; i < conditions.length; i++) {
        if (conditions[i] !== undefined && conditions[i] !== null && !conditions[i].covered())
          message += '\n- '+ conditions[i].message();
    }
    return message;
};

function convertBranchDataConditionArrayToJSON(branchDataConditionArray) {
    var array = [];
    var length = branchDataConditionArray.length;
    for (var condition = 0; condition < length; condition++) {
        var branchDataObject = branchDataConditionArray[condition];
        if (branchDataObject === undefined || branchDataObject === null) {
            value = 'null';
        } else {
            value = branchDataObject.toJSON();
        }
        array.push(value);
    }
    return '[' + array.join(',') + ']';
}

function convertBranchDataLinesToJSON(branchData) {
    if (branchData === undefined) {
        return '[]'
    }
    var array = [];
    var length = branchData.length;
    for (var line = 0; line < length; line++) {
        var branchDataObject = branchData[line];
        if (branchDataObject === undefined || branchDataObject === null) {
            value = 'null';
        } else {
            value = convertBranchDataConditionArrayToJSON(branchDataObject);
        }
        array.push(value);
    }
    return '[' + array.join(',') + ']';
}

function convertBranchDataLinesFromJSON(jsonObject) {
    if (jsonObject === undefined) {
        return [];
    }
    var length = jsonObject.length;
    for (var line = 0; line < length; line++) {
        var branchDataJSON = jsonObject[line];
        if (branchDataJSON !== null) {
            for (var conditionIndex = 0; conditionIndex < branchDataJSON.length; conditionIndex ++) {
                var condition = branchDataJSON[conditionIndex];
                if (condition !== null) {
                    branchDataJSON[conditionIndex] = BranchData.fromJsonObject(condition);
                }
            }
        }
    }
    return jsonObject;
}
try {
  if (typeof top === 'object' && top !== null && typeof top.opener === 'object' && top.opener !== null) {
    // this is a browser window that was opened from another window

    if (! top.opener._$jscoverage) {
      top.opener._$jscoverage = {};
      top.opener._$jscoverage.branchData = {};
    }
  }
}
catch (e) {}

try {
  if (typeof top === 'object' && top !== null) {
    // this is a browser window

    try {
      if (typeof top.opener === 'object' && top.opener !== null && top.opener._$jscoverage) {
        top._$jscoverage = top.opener._$jscoverage;
      }
    }
    catch (e) {}

    if (! top._$jscoverage) {
      top._$jscoverage = {};
      top._$jscoverage.branchData = {};
    }
  }
}
catch (e) {}

try {
  if (typeof top === 'object' && top !== null && top._$jscoverage) {
    this._$jscoverage = top._$jscoverage;
  }
}
catch (e) {}
if (! this._$jscoverage) {
  this._$jscoverage = {};
  this._$jscoverage.branchData = {};
}
if (! _$jscoverage['lib/project.js']) {
  _$jscoverage['lib/project.js'] = [];
  _$jscoverage['lib/project.js'][7] = 0;
  _$jscoverage['lib/project.js'][8] = 0;
  _$jscoverage['lib/project.js'][10] = 0;
  _$jscoverage['lib/project.js'][14] = 0;
  _$jscoverage['lib/project.js'][15] = 0;
  _$jscoverage['lib/project.js'][19] = 0;
  _$jscoverage['lib/project.js'][20] = 0;
  _$jscoverage['lib/project.js'][21] = 0;
  _$jscoverage['lib/project.js'][22] = 0;
  _$jscoverage['lib/project.js'][23] = 0;
  _$jscoverage['lib/project.js'][26] = 0;
  _$jscoverage['lib/project.js'][28] = 0;
  _$jscoverage['lib/project.js'][29] = 0;
  _$jscoverage['lib/project.js'][31] = 0;
  _$jscoverage['lib/project.js'][34] = 0;
  _$jscoverage['lib/project.js'][37] = 0;
  _$jscoverage['lib/project.js'][40] = 0;
  _$jscoverage['lib/project.js'][42] = 0;
  _$jscoverage['lib/project.js'][43] = 0;
  _$jscoverage['lib/project.js'][45] = 0;
  _$jscoverage['lib/project.js'][46] = 0;
  _$jscoverage['lib/project.js'][47] = 0;
  _$jscoverage['lib/project.js'][50] = 0;
  _$jscoverage['lib/project.js'][51] = 0;
}
_$jscoverage['lib/project.js'].source = ["/*!"," * webhook lib/project.js"," * Copyright(c) 2013"," * Author: jifeng &lt;wade428@163.com&gt;"," */","","var fs = require('fs-extra');","var ep = require('event-pipe')","","var util = require('../util');","","","","var Project = function (options) {","  this.options = options;","}","","","Project.prototype.initRep = function (appname, cb) {","\tvar config = this.options","\tvar gitURL = config.git;","\tvar url = gitURL.replace('{appname}', appname);","  var dir = config.rundir + appname;","  ","  ep(function () {","    fs.exists(dir, this);","  }, function (flag) {","  \tif (flag) {","  \t\tfs.removeSync(dir, this);","  \t}","    util.spawn('git', ['clone', url, dir ], this);","  }, function (err, item) {","    //&#24314;&#31435;&#26412;&#22320;stg&#20998;&#25903;","    util.spawn('git', ['checkout', 'stg'], {cwd: dir}, this);","  }, function (err) {","  \t//&#36820;&#22238;master&#20998;&#25903;","    util.spawn('git', ['checkout', 'master'], {cwd: dir}, this);","  }, function (err) {","  \t//&#21028;&#26029;stg&#20998;&#25903;&#30340;&#29256;&#26412;&#26159;&#21542;&#27604;master","  \tutil.spawn('git', ['log',  'stg', '..master'], {cwd: dir}, this);","  }, function (err, info) {","    if (info) {","    \treturn cb(new Error('branch stg is fall behind master\\n' + info));","    }","    var buildFile = dir + '/build.sh'","    util.spawn('sh', [buildFile], {cwd: dir}, cb);","  }).run();","}","","module.exports = function (options) {","\treturn new Project(options);","}"];
_$jscoverage['lib/project.js'][7]++;
var fs = require('fs-extra');
_$jscoverage['lib/project.js'][8]++;
var ep = require('event-pipe');
_$jscoverage['lib/project.js'][10]++;
var util = require('../util');
_$jscoverage['lib/project.js'][14]++;
var Project = function(options) {
  _$jscoverage['lib/project.js'][15]++;
  this.options = options;
};
_$jscoverage['lib/project.js'][19]++;
Project.prototype.initRep = function(appname, cb) {
  _$jscoverage['lib/project.js'][20]++;
  var config = this.options;
  _$jscoverage['lib/project.js'][21]++;
  var gitURL = config.git;
  _$jscoverage['lib/project.js'][22]++;
  var url = gitURL.replace('{appname}', appname);
  _$jscoverage['lib/project.js'][23]++;
  var dir = config.rundir + appname;
  _$jscoverage['lib/project.js'][47]++;
  ep(function() {
  _$jscoverage['lib/project.js'][26]++;
  fs.exists(dir, this);
}, function(flag) {
  _$jscoverage['lib/project.js'][28]++;
  if (flag) {
    _$jscoverage['lib/project.js'][29]++;
    fs.removeSync(dir, this);
  }
  _$jscoverage['lib/project.js'][31]++;
  util.spawn('git', ['clone', url, dir], this);
}, function(err, item) {
  _$jscoverage['lib/project.js'][34]++;
  util.spawn('git', ['checkout', 'stg'], {
  cwd: dir}, this);
}, function(err) {
  _$jscoverage['lib/project.js'][37]++;
  util.spawn('git', ['checkout', 'master'], {
  cwd: dir}, this);
}, function(err) {
  _$jscoverage['lib/project.js'][40]++;
  util.spawn('git', ['log', 'stg', '..master'], {
  cwd: dir}, this);
}, function(err, info) {
  _$jscoverage['lib/project.js'][42]++;
  if (info) {
    _$jscoverage['lib/project.js'][43]++;
    return cb(new Error('branch stg is fall behind master\n' + info));
  }
  _$jscoverage['lib/project.js'][45]++;
  var buildFile = dir + '/build.sh';
  _$jscoverage['lib/project.js'][46]++;
  util.spawn('sh', [buildFile], {
  cwd: dir}, cb);
}).run();
};
_$jscoverage['lib/project.js'][50]++;
module.exports = function(options) {
  _$jscoverage['lib/project.js'][51]++;
  return new Project(options);
};
