// webhook/.cov/run/sqlscan/res/page/ruleproject/ruleproject.js
var Container = require('hc-common/container/container');
var context = require('../../temp/context');
var eventProxy = require('hc-common/core/eventproxy');
var Store = require('hc-common/core/store');
var Msg = require('hc-common/msg/msg');

var RuleProject = function(){};

Container._extend(RuleProject, {
  _className : 'ruleproject',
  __filename: __filename,
  isHide : false,
  initData : function(){
    this.data = {
      url : context.urls
    };
  },
  initEvents : function(){
    this.searchEvent();
    this.rowClickEvent();
    this.tabEvent();
  },
  afterComplete : function(){
    RuleProject.prototype._super.afterComplete.call(this);
    /*var projectname = this.query('combobox[name=projectname]');
    projectname.query = function(searchText){
      return {
        query : JSON.stringify({
          keywords : $.trim(searchText)
        })
      };
    };*/
    this.loadProjectList();
    this.loadProjects();
  },
  searchEvent : function(){
    var _self = this;
    var btnSearch = _self.query('button[name="btnSearch"]');
    btnSearch.on('click', function(){
      _self.loadProjectList();
    });
    /*var btnAddRule = _self.query('button[name="btnAddRule"]');
    btnAddRule.on('click', function(){
      eventProxy.emit('hash.switch', {
        controller : 'ruleadd'
      });
    });*/
  },
  rowClickEvent : function(){
    var _self = this;
    var ruleprojectlist = _self.query('list[name="ruleprojectlist"]');
    ruleprojectlist.on('rowclick', function(record, row, e){
      if($(e.target).attr('class') === 'btnEnableRule'){
        _self.updateProjectRuleByProjectId(record, 'enable');
      }else if($(e.target).attr('class') === 'btnDisableRule'){
        _self.updateProjectRuleByProjectId(record, 'disable');
      }
    })
  },
  tabEvent : function(){
    var _self = this;
    var tab = $('.ruleprojectlist-tab', this.el).children("span");
    tab.on('click' ,function(){
      if($(this).attr('class') === 'ruleprojectlist-tab-active'){
        return;
      }
      for(var i = 0, len = tab.length; i < len; i++){
        if(tab[i] === this){
          tab[i].className = "ruleprojectlist-tab-active"
        }else{
          tab[i].className = "ruleprojectlist-tab-deactive"
        }
      }
      _self.loadProjectList();
    });
  },
  loadProjects : function(){
    var _self = this, userInfo = context.options.session.user;
    var store = Store._create({
      url : context.urls.getProjects,
      param : {
        loginId : userInfo.WorkId, 
        reauestId : '1',
        appId : '',
        status: ''
      }
    });
    store.load(function(err, data){
      if(err){
        return;
      }
      var projectname = _self.query('combobox[name=projectname]');
      projectname.store.load(data);
      _self.emit('afterruleprojectsload');
    });
  },
  setProjectAuthour : function(record, wokid){
    if(record.owner === wokid){
      return true;
    }
    for(var i = 0, roles = record.roles, count = roles.length; i < count; i++){
      if(roles[i].code === "PM"){
        return true;
      }
    }
    return false;
  },
  loadProjectList : function(){
    var _self = this, workId = context.options.session.user.WorkId;
    var ruleprojectlist = this.query('list[name="ruleprojectlist"]');
    var projectId = this.query('combobox[name="projectname"]').val();
    var status = this.getActiveStatus();
    var opt = {
      url : context.urls.getProjects,
      param : {
        loginId : workId, 
        reauestId : '1',
        appId : projectId,
        status: status
      },
      callback : function(){
        //$('#projectcount').html('该条件下共有项目 ' + ruleprojectlist.store.recordCount +' 个');
        _self.emit('afterruleprojectlistload');
      }
    };
    var odpsDisplayName = context.options.session.user.odpsDisplayName;
    ruleprojectlist.beforeRowRender = function(record){
      record.isOwner = _self.setProjectAuthour(record, workId);
    }
    ruleprojectlist.store.load(opt);
  },
  updateProjectRuleByProjectId : function(record, type){
    var _self = this, index = 0, userInfo = context.options.session.user;
    var url = context.urls.enableProjectRuleByProjectId;
    var message = '项目' + record.name + '已启用sqlscan规则';
    if(type === 'disable'){
      url = context.urls.disableProjectRuleByProjectId;
      index = 1;
      message = '项目' + record.name + '已停止sqlscan规则';
    }
    var store = Store._create({
      url : url,
      param : {
        accessId : userInfo.odpsAccessId,
        accessKey : userInfo.odpsRealKey,
        loginId : userInfo.WorkId,
        appId : record.id
      }
    });
    var ruleprojectlist = this.query('list[name="ruleprojectlist"]');
    ruleprojectlist.loading(true);
    store.load(function(err, data){
      ruleprojectlist.loading(false);
      if(err){
        return;
      }
      _self.setActiveStatus(index);
      _self.loadProjectList();
      if(type === 'enable'){
        _self.emit('afterenableprojectrule');
      }else{
        _self.emit('afterdisableprojectrule');
      }
      Msg.alert(message);
    });
  },
  getActiveStatus : function(){
    var tab = $('.ruleprojectlist-tab');
    var status = tab.find('.ruleprojectlist-tab-active').attr('value') || this.activeTab;
    return status;
  },
  setActiveStatus : function(index){
    var tab = $('.ruleprojectlist-tab').children("span");
    for(var i = 0, len = tab.length; i<len ; i++){
      if(i === index){
        tab[i].className = 'ruleprojectlist-tab-active';
      }else{
        tab[i].className = 'ruleprojectlist-tab-deactive';
      }
    }
    //@test
    this.activeTab = index;
  }
});

exports = module.exports = RuleProject;
