// webhook/.cov/util.js
/* ****** automatically generated by jscover - do not edit ******/
if (typeof _$jscoverage === 'undefined') { _$jscoverage = {}; }
/* ****** end - do not edit ******/
function BranchData() {
    this.position = -1;
    this.nodeLength = -1;
    this.src = null;
    this.evalFalse = 0;
    this.evalTrue = 0;

    this.init = function(position, nodeLength, src) {
        this.position = position;
        this.nodeLength = nodeLength;
        this.src = src;
        return this;
    }

    this.ranCondition = function(result) {
        if (result)
            this.evalTrue++;
        else
            this.evalFalse++;
    };

    this.pathsCovered = function() {
        var paths = 0;
        if (this.evalTrue > 0)
          paths++;
        if (this.evalFalse > 0)
          paths++;
        return paths;
    };

    this.covered = function() {
        return this.evalTrue > 0 && this.evalFalse > 0;
    };

    this.toJSON = function() {
        return '{"position":' + this.position
            + ',"nodeLength":' + this.nodeLength
            + ',"src":' + jscoverage_quote(this.src)
            + ',"evalFalse":' + this.evalFalse
            + ',"evalTrue":' + this.evalTrue + '}';
    };

    this.message = function() {
        if (this.evalTrue === 0 && this.evalFalse === 0)
            return 'Condition never evaluated         :\t' + this.src;
        else if (this.evalTrue === 0)
            return 'Condition never evaluated to true :\t' + this.src;
        else if (this.evalFalse === 0)
            return 'Condition never evaluated to false:\t' + this.src;
        else
            return 'Condition covered';
    };
}

BranchData.fromJson = function(jsonString) {
    var json = eval('(' + jsonString + ')');
    var branchData = new BranchData();
    branchData.init(json.position, json.nodeLength, json.src);
    branchData.evalFalse = json.evalFalse;
    branchData.evalTrue = json.evalTrue;
    return branchData;
};

BranchData.fromJsonObject = function(json) {
    var branchData = new BranchData();
    branchData.init(json.position, json.nodeLength, json.src);
    branchData.evalFalse = json.evalFalse;
    branchData.evalTrue = json.evalTrue;
    return branchData;
};

function buildBranchMessage(conditions) {
    var message = 'The following was not covered:';
    for (var i = 0; i < conditions.length; i++) {
        if (conditions[i] !== undefined && conditions[i] !== null && !conditions[i].covered())
          message += '\n- '+ conditions[i].message();
    }
    return message;
};

function convertBranchDataConditionArrayToJSON(branchDataConditionArray) {
    var array = [];
    var length = branchDataConditionArray.length;
    for (var condition = 0; condition < length; condition++) {
        var branchDataObject = branchDataConditionArray[condition];
        if (branchDataObject === undefined || branchDataObject === null) {
            value = 'null';
        } else {
            value = branchDataObject.toJSON();
        }
        array.push(value);
    }
    return '[' + array.join(',') + ']';
}

function convertBranchDataLinesToJSON(branchData) {
    if (branchData === undefined) {
        return '[]'
    }
    var array = [];
    var length = branchData.length;
    for (var line = 0; line < length; line++) {
        var branchDataObject = branchData[line];
        if (branchDataObject === undefined || branchDataObject === null) {
            value = 'null';
        } else {
            value = convertBranchDataConditionArrayToJSON(branchDataObject);
        }
        array.push(value);
    }
    return '[' + array.join(',') + ']';
}

function convertBranchDataLinesFromJSON(jsonObject) {
    if (jsonObject === undefined) {
        return [];
    }
    var length = jsonObject.length;
    for (var line = 0; line < length; line++) {
        var branchDataJSON = jsonObject[line];
        if (branchDataJSON !== null) {
            for (var conditionIndex = 0; conditionIndex < branchDataJSON.length; conditionIndex ++) {
                var condition = branchDataJSON[conditionIndex];
                if (condition !== null) {
                    branchDataJSON[conditionIndex] = BranchData.fromJsonObject(condition);
                }
            }
        }
    }
    return jsonObject;
}
try {
  if (typeof top === 'object' && top !== null && typeof top.opener === 'object' && top.opener !== null) {
    // this is a browser window that was opened from another window

    if (! top.opener._$jscoverage) {
      top.opener._$jscoverage = {};
      top.opener._$jscoverage.branchData = {};
    }
  }
}
catch (e) {}

try {
  if (typeof top === 'object' && top !== null) {
    // this is a browser window

    try {
      if (typeof top.opener === 'object' && top.opener !== null && top.opener._$jscoverage) {
        top._$jscoverage = top.opener._$jscoverage;
      }
    }
    catch (e) {}

    if (! top._$jscoverage) {
      top._$jscoverage = {};
      top._$jscoverage.branchData = {};
    }
  }
}
catch (e) {}

try {
  if (typeof top === 'object' && top !== null && top._$jscoverage) {
    this._$jscoverage = top._$jscoverage;
  }
}
catch (e) {}
if (! this._$jscoverage) {
  this._$jscoverage = {};
  this._$jscoverage.branchData = {};
}
if (! _$jscoverage['util.js']) {
  _$jscoverage['util.js'] = [];
  _$jscoverage['util.js'][7] = 0;
  _$jscoverage['util.js'][8] = 0;
  _$jscoverage['util.js'][9] = 0;
  _$jscoverage['util.js'][11] = 0;
  _$jscoverage['util.js'][12] = 0;
  _$jscoverage['util.js'][13] = 0;
  _$jscoverage['util.js'][14] = 0;
  _$jscoverage['util.js'][16] = 0;
  _$jscoverage['util.js'][17] = 0;
  _$jscoverage['util.js'][18] = 0;
  _$jscoverage['util.js'][19] = 0;
  _$jscoverage['util.js'][20] = 0;
  _$jscoverage['util.js'][21] = 0;
  _$jscoverage['util.js'][23] = 0;
  _$jscoverage['util.js'][27] = 0;
  _$jscoverage['util.js'][31] = 0;
  _$jscoverage['util.js'][32] = 0;
  _$jscoverage['util.js'][33] = 0;
  _$jscoverage['util.js'][34] = 0;
  _$jscoverage['util.js'][36] = 0;
  _$jscoverage['util.js'][37] = 0;
  _$jscoverage['util.js'][38] = 0;
  _$jscoverage['util.js'][39] = 0;
  _$jscoverage['util.js'][40] = 0;
  _$jscoverage['util.js'][41] = 0;
  _$jscoverage['util.js'][42] = 0;
  _$jscoverage['util.js'][44] = 0;
  _$jscoverage['util.js'][45] = 0;
  _$jscoverage['util.js'][46] = 0;
  _$jscoverage['util.js'][49] = 0;
  _$jscoverage['util.js'][51] = 0;
  _$jscoverage['util.js'][52] = 0;
  _$jscoverage['util.js'][53] = 0;
  _$jscoverage['util.js'][54] = 0;
  _$jscoverage['util.js'][57] = 0;
  _$jscoverage['util.js'][58] = 0;
}
_$jscoverage['util.js'].source = ["/*!"," * webhook util.js"," * Copyright(c) 2013"," * Author: jifeng &lt;wade428@163.com&gt;"," */","","var fs = require('fs');","var spawn = require('child_process').spawn;","var os = require('options-stream');","","var getUidAndGid = function () {","  var uid = process.getuid();","  if (uid &gt;= 500) {","    return {uid: process.getuid(), gid: process.getgid()}","  }","  var gidFile = gid_file || '/etc/passwd';","  var str = fs.readFileSync(gidFile, 'utf-8');","  reg  = /[^app]admin:x:+(\\d+):(\\d+)/","  res  = str.match(reg);","  if (!res) {","    return","  }","  var user = {","    uid : +res[1],","    gid : +res[2]","  }","  return user;","}","","","exports.spawn = function (command, args, options, cb) {","  if ('function' === typeof options) {","    cb = options;","    options = {};","  }","  var user = getUidAndGid();","  user = os(user, options);","  var chunks = [];","  var error = false;","  proc = spawn(command, args, user);","  proc.stdout.on('data', function (data) {","    chunks.push(data);","  });","  proc.stderr.on('data', function (data) {","    error = true;","    chunks.push(data);","  });","","  proc.on('close', function (code) {","    //&#36816;&#34892;&#26576;&#20123;&#25351;&#20196;&#32791;&#26102;&#27604;&#36739;&#38271;&#65292;&#25152;&#20197;&#35774;&#32622;&#30340;callback&#24310;&#26102;","    var str = Buffer.concat(chunks).toString();","    if (error) {","      return setTimeout(function () {","        cb (new Error(str));","      }, 50);","    }","    setTimeout(function () {","      cb(null, str);","    }, 100);","    ","  });","}"];
_$jscoverage['util.js'][7]++;
var fs = require('fs');
_$jscoverage['util.js'][8]++;
var spawn = require('child_process').spawn;
_$jscoverage['util.js'][9]++;
var os = require('options-stream');
_$jscoverage['util.js'][11]++;
var getUidAndGid = function() {
  _$jscoverage['util.js'][12]++;
  var uid = process.getuid();
  _$jscoverage['util.js'][13]++;
  if (uid >= 500) {
    _$jscoverage['util.js'][14]++;
    return {
  uid: process.getuid(), 
  gid: process.getgid()};
  }
  _$jscoverage['util.js'][16]++;
  var gidFile = gid_file || '/etc/passwd';
  _$jscoverage['util.js'][17]++;
  var str = fs.readFileSync(gidFile, 'utf-8');
  _$jscoverage['util.js'][18]++;
  reg = /[^app]admin:x:+(\d+):(\d+)/;
  _$jscoverage['util.js'][19]++;
  res = str.match(reg);
  _$jscoverage['util.js'][20]++;
  if (!res) {
    _$jscoverage['util.js'][21]++;
    return;
  }
  _$jscoverage['util.js'][23]++;
  var user = {
  uid: +res[1], 
  gid: +res[2]};
  _$jscoverage['util.js'][27]++;
  return user;
};
_$jscoverage['util.js'][31]++;
exports.spawn = function(command, args, options, cb) {
  _$jscoverage['util.js'][32]++;
  if ('function' === typeof options) {
    _$jscoverage['util.js'][33]++;
    cb = options;
    _$jscoverage['util.js'][34]++;
    options = {};
  }
  _$jscoverage['util.js'][36]++;
  var user = getUidAndGid();
  _$jscoverage['util.js'][37]++;
  user = os(user, options);
  _$jscoverage['util.js'][38]++;
  var chunks = [];
  _$jscoverage['util.js'][39]++;
  var error = false;
  _$jscoverage['util.js'][40]++;
  proc = spawn(command, args, user);
  _$jscoverage['util.js'][41]++;
  proc.stdout.on('data', function(data) {
  _$jscoverage['util.js'][42]++;
  chunks.push(data);
});
  _$jscoverage['util.js'][44]++;
  proc.stderr.on('data', function(data) {
  _$jscoverage['util.js'][45]++;
  error = true;
  _$jscoverage['util.js'][46]++;
  chunks.push(data);
});
  _$jscoverage['util.js'][49]++;
  proc.on('close', function(code) {
  _$jscoverage['util.js'][51]++;
  var str = Buffer.concat(chunks).toString();
  _$jscoverage['util.js'][52]++;
  if (error) {
    _$jscoverage['util.js'][53]++;
    return setTimeout(function() {
  _$jscoverage['util.js'][54]++;
  cb(new Error(str));
}, 50);
  }
  _$jscoverage['util.js'][57]++;
  setTimeout(function() {
  _$jscoverage['util.js'][58]++;
  cb(null, str);
}, 100);
});
};
