// webhook/run/sqlscan/lib/app-sqlscan.js
var Emitter = require('events').EventEmitter;
var util = require('util');

var justlog = require('justlog');
var tfs = require('tfs');
var queryString = require('querystring');

var SqlScan = function (options) {
  options = options || {};
  this.options = options;
  this.client = tfs.createClient({
    rootServer: options.prefix,
    appkey: options.appkey || '51fb509427789'
  });
  if (options.logs_dir) {
    this.log = justlog({
      file: {
        level : justlog.ERROR | justlog.INFO,
        path : "[" + options.logs_dir + "/sqlscan-]YYYY-MM-DD[.log]",
        pattern : 'file'
      },
      stdio: false
    });
  }
};

util.inherits(SqlScan, Emitter);

SqlScan.prototype.upload = function (contents, cb) {
  contents = new Buffer(contents, 'base64').toString();
  var buffer = '';
  try{
    buffer = new Buffer(contents);
  }catch(err){
    return cb(err);
  }
  this.client.upload(buffer, '', function(err, data){
    cb(err, data);
  });
}

SqlScan.prototype._get = function (req, res, cb) {
  var params = req.url.split('/');
  if(params.length !== 3){
    cb('url error， fileContents needed');
    return;
  }
  var contents = queryString.unescape(params[2]);
  cb(null, contents);
}

SqlScan.prototype._post = function (req, res, cb) {
  var chunks = [];
  req.on('data', function(chunk){
    chunks.push(chunk);
  });
  req.on('error', function(err){
    cb(err);
  });
  req.on('end', function(){
    chunks = queryString.unescape(Buffer.concat(chunks).toString());
    //store里面会默认给data添加'/'
    chunks = chunks.substr(1);
    cb(null, chunks);
  });
}

SqlScan.prototype.request = function (req, res, cb) {
  var method = req.method.toLowerCase();
  if(method === 'post' || method === 'put') {
    this._post(req, res, cb);
  }else{
    this._get(req, res, cb);
  }
}

SqlScan.prototype.middleware = function() {
  var _self = this;
  return function (req, res, next) {
    var url = req.url, requestUrl;
    //var user = req.session && req.session.user && req.session.user.nick;
    if (0 === url.indexOf('/upload')) {
      requestUrl = '/upload/';
    }
    if (!requestUrl) {
      res.statusCode = 404;
      return res.end();
    }
    _self.request(req, res, function(err, contents){
      if (err) {
        if (_self.log) {
          _self.log.error('REQUEST SqlScan ERROR: ' + err.toString());
        }
        return res.end(JSON.stringify({returnCode: '1', returnMessage: err.toString()}));
      }
      _self.upload(contents, function (err, data) {
        if (err) {
          res.statusCode = 500;
          if (_self.log) {
            _self.log.error('REQUEST SqlScan ERROR: ' + err.toString());
          }
          return res.end(JSON.stringify({returnCode: '1', returnMessage: err.toString()}));
        }
        return res.end(JSON.stringify({returnCode: '0', returnValue: data}));
      });
    }); 
  };
};

module.exports = function (options) {
  return new SqlScan(options);
};
