// webhook/run/sqlscan/res/page/index/viewport.js
var App = require('../app');
var $ = require('jquery');
var context = require('../../temp/context');

/**
 * 应用配置项
 */
var config = {
  coms : {
    /*toolbar : {
      com: '/module/page/common/toolbar/index.js',
      autoload : true,
      config : {
        renderTo : $('.header'),
        holddisplay : true,
        isHelperHide: true
      }
    },*/
    fnmenu : {
      com : '/page/common/fnmenu/fnmenu.js',
      autoload : true,
      config : {
        renderTo : $('.left-menu'),
        param : {
          'module' : 'sqlscan',
          'loadRullAdd' : 'ruleadd'
        },
        holddisplay : true
      }
    },
    ruleview : {
      com : '/page/ruleview/ruleview.js',
      config : {
        renderTo : $('.workspace-body'),
        autoDestroy : true
      }
    },
    rulemanage : {
      com : '/page/rulemanage/rulemanage.js',
      config : {
        renderTo : $('.workspace-body'),
        autoDestroy : true
      }
    },
    ruleadd : {
      com : '/page/ruleadd/ruleadd.js',
      config : {
        renderTo : $('.workspace-body'),
        autoDestroy : true
      }
    },
    project : {
      com : '/page/ruleproject/ruleproject.js',
      config : {
        renderTo : $('.workspace-body'),
        autoDestroy : true
      }
    }
  }
};
/**
 * 初始化
 */
function ViewPort(opt){
  root.ajaxGet('/s/sess', function(err, data){
    opt['session'] = JSON.parse(data) || {};
    root.ajaxGet('/r/apps/user/getUserOdpsInfo', function(err, newData){
      var rtn = JSON.parse(newData) || {};
      if(rtn.returnCode == 0){
        opt['session'].user.odpsRealKey = rtn.returnValue.odpsAccessKey;
      }
      context.init(opt);
      var app = App._create(config);
      if(window.history.pushState){
        $(window).bind('popstate',function(e){
          if(e.originalEvent.state){
            app.onRouterChange();
          }
        });
      }else{
        $(window).bind('hashchange',function(e){
          app.onRouterChange();
        });
      }
    });
  });
}
exports = module.exports = ViewPort;
