// webhook/.cov/run/sqlscan/res/page/ruleadd/ruleadd.js
var Container = require('hc-common/container/container');
var context = require('../../temp/context');
var AuthorizeItems = require('./authorizeitem/authorizeitem');
var Store = require('hc-common/core/store');
var eventProxy = require('hc-common/core/eventproxy');
var Msg = require('hc-common/msg/msg');

var Ruleadd = function(){};

Container._extend(Ruleadd, {
  _className : 'ruleadd',
  __filename: __filename,
  isHide : false,
  auhtorItems : 0,
  initData : function(){
    this.data = {
      url : context.urls,
      ruleInfo : {
        users : {},
        ruleBean : {}
      }
    };
    if(this.router.ruleInfo){
      this.router.ruleInfo = JSON.parse(this.router.ruleInfo);
      this.data.ruleInfo = this.router.ruleInfo;
    }
  },
  initEvents : function(){
    this.initResourceEvent();
    this.initAuthorizeRuleEvent();
    this.initAddEvent();
  },
  afterComplete : function(){
    Ruleadd.prototype._super.afterComplete.call(this);
    if(this.router.ruleInfo){//label标签是html标签，不支持jade
      $('#sqlScanResourceId').attr("resourceId", this.router.ruleInfo.ruleBean.resourceId);
    }
  },
  initResourceEvent : function(){
    var _self = this;
    var btnSource = _self.query('button[name=btnSource]');
    btnSource.on('click', function(){
      root.loadJs('/page/selectresource/selectresource.js', function(err, objs){
        var win = objs['/page/selectresource/selectresource.js']._create({
          title : '选择资源'
        });
        _self.emit('showselectresourcewin');
      });
    });
    var btnCreate = _self.query('button[name=btnCreate]');
    btnCreate.on('click', function(){
      root.loadJs('/page/createresource/createresource.js', function(err, objs){
        var win = objs['/page/createresource/createresource.js']._create({
          title : '创建资源'
        });
        _self.emit('showcreateresourcewin');
      });
    });
  },
  initAuthorizeRuleEvent : function(){
    var _self = this;
    var btnAddAuthorize = _self.query('button[name="btnAddAuthorize"]');
    btnAddAuthorize.on('click', function(){
      _self.addAuthorizeRule();
    });
  },
  initAddEvent : function(){
    var _self = this;
    var btnAdd = this.query('button[name=btnAdd]');
    btnAdd.on('click', function(){
      _self.createRuleAction();
    });
    var btnCancel = this.query('button[name=btnCancel]');
    btnCancel.on('click', function(){
      _self.gotoRuleManage();
    });
  },
  addAuthorizeRule : function(){
    var itemDom = $('.authorizecontainer', this.el);
    var opt = {
      ctype: AuthorizeItems,
      name: 'authorizeitems',
      renderTo: itemDom
    };
    this.addItem(opt);
    this.auhtorItems ++;
  },
  gotoRuleManage : function(){
    eventProxy.emit('menu.switch', {
      controller : 'rulemanage'
    });
    this.testCheckValue = '#controller:rulemanage';
  },
  createRuleAction : function(){
    var _self = this, userInfo = context.options.session.user;
    var name = this.query('field[name=rulename]');
    var resource = $('#sqlScanResourceId');
    var desc = this.query('textareafield[name=ruledesc]');
    var authorizeInfosArr = this.getAuthorizeInfos(userInfo.nickNameCn);
    if(!name.validate() || authorizeInfosArr === 'error' || !desc.validate()){
      _self.testCheckValue = '参数不正确';
      return;
    }else if(!resource.attr("resourceId") && !_self.resourceIdTest){
      Msg.alert('资源不能为空');
      _self.testCheckValue = '资源不能为空';
      return;
    }
    var rank = $("input[name='ruleRank']:checked").val();
    var ruleVariable = $("input[name='ruleVariable']:checked").val();
    var url = context.urls.insertRuleAndConfigure;
    var message = '规则添加成功';
    var ruleParams = {
      name : $.trim(name.val()),
      rank : parseInt(rank),
      replaceParamStatus : parseInt(ruleVariable),
      resourceId : parseInt(resource.attr("resourceId")),
      detail : $.trim(desc.val()),
      author : userInfo.nickNameCn,
      authorUserName : userInfo.odpsDisplayName
    };
    if(this.router.ruleInfo){//编辑规则需要ruleId
      url = context.urls.updateRule;
      message = '规则编辑成功';
      ruleParams.id = this.router.ruleInfo.ruleBean.id;
    }
    var store = Store._create({
      url : url,
      param : {
        ruleInfos : JSON.stringify(ruleParams),
        authorizeInfos : authorizeInfosArr,
        access_id : userInfo.odpsAccessId,
        access_key : userInfo.odpsRealKey,
        work_id : userInfo.userId
      }
    });
    store.load(function(err, data){
      if(err){
        return;
      }
      _self.gotoRuleManage();
      _self.testCheckValue = message;
      _self.emit('addrulesucess');
      Msg.alert(message);
    });
  },
  getAuthorizeItemInfo : function(appId, owner, author){
    if(!appId.validate() || !owner.validate()){
      return 'error';
    }
    var appRecord = appId._getRecord(appId.val());
    var userName = {};
    for(var i = 0, items = owner.val(), len = items.length; i < len; i++){
      if(items[i] === '*'){
        userName['*'] = '*';
        break;
      }
      var ownerRecord = owner._getRecord(items[i]);
      userName[ownerRecord.odpsDisplayName] = ownerRecord.nickName;
    }
    var userInfo = context.options.session.user;
    var rtn = JSON.stringify({
      className : "com.taobao.dw.sqlscan.dto.ConfigureDto",
      properties : {
        projectName : appRecord.odpsProjectName,
        projectRealName : appRecord.name,
        userName : userName,
        author : author,
        authorUserName : userInfo.odpsDisplayName
      }
    });
    return rtn;
  },
  getAuthorizeInfos : function(author){
    var arr = [];/*, authorizeitemIndex = $('.authorizecontainer', this.el).children().length*/
    var apps = this.query('combobox[name=appId]');
    var owners = this.query('combobox[name=owner]');
    for(var i = 0, len = apps.length; i < len; i++){
      var item = this.getAuthorizeItemInfo(apps[i], owners[i], author);
      if(item === 'error'){
        return 'error';
      }
      arr.push(item);
    }
    return arr;
  },
  query : function(selector){
    var i, len;
    var comp = Container.prototype._super.query.call(this, selector);
    if(comp){
      return comp;
    }else{
      var rtn = [];
      for(i = 0, len = this.comps.length; i < len; i++){
        if(this.comps[i]._classType === 'container'){
          comp = this.comps[i].query(selector);
          if(comp && this.comps[i].el){
            rtn.push(comp);
          }
        }
      }
      return rtn;
    }
    return null;
  }
});

exports = module.exports = Ruleadd;
