// webhook/run/sqlscan/res/page/common/feedback/feedback.js
var WindowPanel = require('../../../com/window/windowpanel');
var context = require('../context');
var Msg = require('../../../com/msg/msg');
var Store = require('../../../core/store');
var IssueList = require('../issuelist/issuelist');

var FeedBack = function(){};
WindowPanel._extend(FeedBack, {
  _className : 'FeedBack',
  __filename : __filename,
  width : '480px',
  height : '325px',
  tplName : 'page/common/feedback/feedback',
  cssName : 'page/common/feedback/feedback',
  userInfo : null,
  prefix : '',
  afterComplete : function(){
    FeedBack.prototype._super.afterComplete.call(this);
    var _self = this;
    $('.feed_back_fixed').addClass('feed_back_show');
    $(_self.el).css('position', 'fixed');
  },
  initEvents : function(){
    FeedBack.prototype._super.initEvents.call(this);
    var _self     = this;
    var feedbackFixDiv = $('.feed_back_fixed');
    var issueList = $('.his-issue-list', this.el);
    var submitBtn = _self.query('button[name="doSubmitBtn"]');
    var titleSpan = $('.issue-title span', this.el);
    var cancelBtn = _self.query('button[name="doCancelBtn"]');

    feedbackFixDiv.on('click', function(){
      var feedbackClazz = feedbackFixDiv.attr('class');
      if (feedbackClazz.indexOf('feed_back_show') > -1){
        (feedbackFixDiv.removeClass('feed_back_show')).addClass('feed_back_hide');
        $(_self.el).hide();
      } else if (feedbackClazz.indexOf('feed_back_hide') > -1){
        (feedbackFixDiv.removeClass('feed_back_hide')).addClass('feed_back_show');
        $(_self.el).show();
      }
    }); 
    //for test
    _self.on('addissuereturn', function(data){
      _self.doSubmitReturn(data);
    })
    _self.on('issuelistreturn', function(data){
      _self.getIssueListReturn(data);
    })
    //end for test
    $('body').on('click.feedback',function(e){
      if (($(e.target).attr('class') && $(e.target).hasClass('feed_back_fixed') == true) || $(e.target).parents('.window-feedback').length > 0){
        return;
      }
      _self.hideSelf();
    });
    submitBtn.on('click', _self.doSubmit.bind(_self));
    issueList.on('click', _self.getIssueList.bind(_self));
    titleSpan.on('click', function(){
      titleSpan.removeClass('active-title');
      $(this).addClass('active-title');
    });
    cancelBtn.on('click', function(){
      _self.hideSelf();
    });
    $('textarea', _self.el).on('keyup',_self.checkInput);
  },
  checkInput : function(){
    var _self = this;
    var textareaContent = $('textarea', _self.el);
    if (textareaContent.val()){
      textareaContent.css('border-color', 'rgba(0, 0, 0, 0.3)');
      return true;
    } else {
      textareaContent.css('border-color', 'red');
      return false;
    }
  },
  doSubmit : function(){
    var _self = this;
    _self.query('button[name="doSubmitBtn"]').disabled();
    if (_self.checkInput()){
      var title       = $('.issue-title .active-title', _self.el).text();
      var description = $('textarea', _self.el).val();
      var label       = $('input[type="radio"]:checked', _self.el).val();
      var nickname = _self.userInfo.nickNameCn || _self.userInfo.nick || _self.userInfo.WorkId;
      url = '/feedback/addIssue';
      if (_self.prefix != ''){
        url = _self.prefix + url;
      }
      $.ajax({
        type : 'GET',
        url : url,
        data : {
          'title'       : nickname + '&nbsp;' + title + ':' + description.substring(0, 40),
          'description' : description,
          'labels'      : label
        },
        dataType : 'jsonp',
        success : function(data){
          _self.doSubmitReturn(data);
        }
      });
    } else {
      _self.query('button[name="doSubmitBtn"]').enabled();
      $('textarea', _self.el).focus().css('border-color', 'red');
    }
  },
  doSubmitReturn : function(data){
    var _self = this;
    var obj = new Function( 'return (' + data + ')') ();
    if (obj.returnCode == 'mycode'){
      Msg.alert('您还未注册过GitLab，请先登陆<a href="http://gitlab.alibaba-inc.com" target="_blank">http://gitlab.alibaba-inc.com</a>');
    } else {
      if (obj.id){
        $('.issue-title', _self.el).hide();
        $('.submit-btn', _self.el).hide();
        _self.query('button[name="doCancelBtn"]').val('确定');
        var issueBody = $('.issue-body', _self.el);
        var issueBodyChildren = issueBody.children();
        var div = $('<div class="create-issue-return"><span>反馈已收到，我们会尽快处理：）</span></div>');
        issueBody.html(div);
        setTimeout(function(){
          $('.issue-title', _self.el).show().find('span').removeClass('active-title').first().addClass('active-title');
          $('.submit-btn', _self.el).show();
          _self.query('button[name="doCancelBtn"]').val('取消');
          issueBody.html(issueBodyChildren);
          $('textarea', _self.el).val('');
          $('input[type="radio"]', _self.el)[2].checked = true;
          _self.hideSelf();
          _self.emit('endcreateissue');
        }, 3000);
      } else {
        $('textarea', _self.el).val('').attr('placeholder', '添加失败').css('border-color', 'red');
      }
    }
    _self.query('button[name="doSubmitBtn"]').enabled();
  },
  getIssueList : function(){
    var _self = this;
    var url = '/feedback/getIssueList';
    if (_self.prefix != ''){
      url = _self.prefix + url;
    }
    $.ajax({
      type : 'GET',
      url : url,
      dataType : 'jsonp',
      success : function(data){
        _self.getIssueListReturn(data);
      }
    });
  },
  getIssueListReturn : function(data){
    var data = new Function( 'return (' + data + ')') ();
    var top = $('.feed_back_fixed').css('top');
    top = top ? top.substring(0, top.length - 2) : 0;
    if (data != null || data.length > 0){
      IssueList._create({
        record : data,
        drag : false,
        modal : false,
        top : Math.max(top - 80, 0),
        left : document.body.clientWidth - 530
      });
    }
    this.hideSelf();
  },
  hideSelf : function(){
    var feedbackClazz = $('.feed_back_fixed').attr('class');
    ($('.feed_back_fixed').removeClass('feed_back_show')).addClass('feed_back_hide');
    $(this.el).hide();
  }
});
exports = module.exports = FeedBack;
