// webhook/run/sqlscan/res/page/common/navs/index.js
var $ = require('jquery');
var Container = require('../../../com/container/container');
var eventProxy = require('../../../core/eventproxy');
var Buttons = require('../../../newcom/button/index');
var Tip = require('../../../newcom/tip/index');
var context = require('../context');

var Index = function(){};

Container._extend(Index, {
  _className : 'index',
  __filename: __filename,
  navflag: 'false',
  prefix: '',
  _loadStore: function(done){
    var _self = this;
    this._getToolsInfo(function(err, data){
      if(!err){
        _self.myTools = data;
        done();
      }
    });
  },
  afterComplete: function(){
    var _self = this;
    Index.prototype._super.afterComplete.call(this);
    this.pluginSetting = this.query("*[name='pluginSetting']");
    this.btnSetting = $('.btn', _self.pluginSetting.el);
    this.navbody = $('.navbody', this.el);
    this.settingContainer = $('.setting', this.el);
    eventProxy.on('tools.destroy', function(){
      _self.propoverDestroy();
      _self.btnSetting.trigger('click');
    });
    //动态变更当前工具栏
    this.toolsChange();
    _self.on('cwf.env.change', function(env){
      var cwf = $('.ticon-cwf', this.navbody);
      var cwfdev = $('.ticon-cwfdev', this.navbody);
      if(env === 'pre'){
        cwfdev.addClass('cwfdevsel btnsel');
        cwfdev.parent().addClass('selected');
        cwf.removeClass('cwfsel').removeClass('btnsel');
        cwf.parent().removeClass('selected');
      }else{
        cwf.addClass('cwfsel btnsel');
        cwf.parent().addClass('selected');
        cwfdev.removeClass('cwfdevsel').removeClass('btnsel');
        cwfdev.parent().removeClass('selected');
      }
    });

    // this.initSelNav();
  },
  toolsChange: function(){
    var len, i, itemDom, item, opt, _self = this, soffset, swidth;
    eventProxy.on('tools.change', function(ids, tools){
      _self.navbody.find('.item').remove();
      len = tools ? tools.length : 0;
      if(len >= 1){
        for(i=0; i < len; i ++){
          itemDom = $("<div class='item'>");
          itemDom.appendTo(_self.navbody).insertBefore(_self.settingContainer);
          item = tools[i];
          if(item){
            var btnContainer = $("<div class='btnContainer'></div>").appendTo(itemDom);
            var newA = $("<a></a>").appendTo(btnContainer);
            newA.attr('name', item.alias);
            newA.attr('class', 'toolbaricon common-btn ticon-' + item.alias);
            newA.attr('datatitles', item.name);
            newA.attr('href', item.path);
            newA.attr('dataurl', item.path);
            newA.attr('target', '_blank');
          }
        }
        _self.bindTip();
        _self.initSelNav();
      }
      soffset = _self.pluginSetting.el.offset();
      swidth = _self.pluginSetting.el.outerWidth(true);
      eventProxy.emit('nav.setting.offset', soffset, swidth);
    });
  },
  initData: function(){
    var _self = this;
    _self.data = {
      myTools: _self.myTools
    };
  },
  initEvents: function(){
    Index.prototype._super.initEvents.call(this);
    this._settingEvent();
    //绑定tip
    this.bindTip();
    this.initSelNav();
  },
  _settingEvent: function(){
    var _self = this;
    var left, minLeft, popoverWidth, bodyWidth, right, popover = null, opt, tcontiner, swidth, soffset;
    tcontiner = $('.mininav .container');
    var pluginfn = function(){
      var popoverWidth;
      var popover = require('../../../newcom/popover/index.js')._create({
        events: {
          'aftercomplete': function(){
            var popoverBody = $('.popover-content', popover.el);
            _self.ps = require('../pluginsetting/index.js')._create({
              renderTo: popoverBody,
              prefix: _self.prefix,
              events: {
                'aftercomplete' : function(){
                  if(tcontiner.offset()){
                    left = tcontiner.offset().left;
                  }
                  bodyWidth = $(document.body).outerWidth(true);
                  right = left;
                  var appLeft, _width = $('.mininav .container').width();
                  var $app = $('.mininav .app');
                  if(_self.navflag === 'true'){
                    appLeft = $app.offset()!== undefined ? $app.offset().left : 0;
                    appLeft += undefined !== $app ? $app.width() : 0;
                    opt = {
                      top : soffset.top + _self.pluginSetting.el.outerHeight(true) + 10,
                      left: appLeft
                    };
                  }else{
                    opt = {
                      top : soffset.top + _self.pluginSetting.el.outerHeight(true) + 10,
                      right: right
                    };
                  }
                  popover.show(opt);
                  popoverWidth = popoverBody.width();
                  if(_self.navflag === 'true'){
                    minLeft = appLeft;
                  }else{
                    minLeft = bodyWidth - popoverWidth - right;
                  }
                  popover.setArrowPosition(soffset.left - minLeft + swidth /2);
                  eventProxy.on('nav.setting.offset', function(offset, width){
                    popover.setArrowPosition(offset.left - minLeft + width/2);
                  });
                }
              }
            });
          }
        }
      });
      return popover;
    };
    this.pluginSetting.on('click', function(){
      soffset = _self.pluginSetting.el.offset();
      swidth = _self.pluginSetting.el.outerWidth(true);
      if(_self.btnSetting.hasClass('selplugin')){
        _self.propoverDestroy();
      }else{
        _self.propoverDropdown();
        _self.popover = pluginfn();
        $(window).bind('click.dropup', function(e){
          if (_self.btnSetting.hasClass('selplugin') && !$.contains(_self.el[0], e.target) && !$.contains($('.popover-content')[0], e.target)){
            if(_self.ps && _self.ps.settingFlag){
              _self.ps.saveMyTools(function(flag){
                if(flag){
                  _self.propoverDestroy();
                  _self.btnSetting.removeClass('selplugin');
                  _self.ps = null;
                }
              });
            }else{
              _self.propoverDestroy();
              _self.btnSetting.removeClass('selplugin');
              _self.ps = null;
            }
          }
        });
      }
    });
    this.pluginSetting.el.find("button.js-dropdown").on('dropup', function(){
      _self.propoverDestroy();
      _self.btnSetting.removeClass('selplugin');
    });
  },
  _getToolsInfo: function(cb){
    var url = this.prefix + '/nr/header/view';
    $.ajax({
      type: 'GET',
      url: url,
      dataType:'jsonp',
      async: false,
      success: function(data){
        if(data){
          data = data.replace(/\n|\r/g, '');
          data = data.replace(/\\/g, '\\\\');
          var obj = JSON.parse(data);
          cb(null, obj.myTools);
        }
      },
      error: function(jqXHR, textStatus, errorThrown){
        cb(errorThrown);
      }
    });
  },
  _getTools: function(cb){
    var _self = this;
    var url = this.prefix + '/nr/header/tools';
    if(localStorage.getItem('global_toolbar_all_tools')){
      cb(null, _self.parseString(localStorage.getItem('global_toolbar_all_tools')));
    }else{
      $.ajax({
        type: 'GET',
        url: url,
        dataType:'jsonp',
        async: false,
        success: function(data){
          localStorage.setItem('global_toolbar_all_tools', data);
          cb(null, _self.parseString(data));
        },
        error: function(jqXHR, textStatus, errorThrown){
          cb(errorThrown);
        }
      });
    }
  },
  parseString: function(str){
    return (new Function( 'return (' + str + ')') ());
  },
  propoverDropdown: function(){
    $('.loginAll button.js-dropdown').trigger('dropup');
  },
  propoverDestroy: function(){
    var _self = this;
    if(_self.popover){
      _self.popover.destroy();
      _self.popover = null;
    }
    $(window).unbind('click.dropup');
  },
  bindTip: function(){
    var _self = this;
    $('.toolbaricon', this.navbody).each(function(index, item){
      //选中当前菜单
      var pathname = window.location.pathname;
      var hostname = window.location.hostname;
      _self.selectNav(pathname, hostname, item);
      _self.navBind(item);
    });
  },
  navBind: function(item){
    var _self = this;
    //替换url{appId}
    _self.beautifyUrl(item);
    //绑定mouseenter & mouseleave
    $(item).on('mouseenter', function(){
      _self.showTip($(item));
    });
    $(item).on('mouseleave', function(){
      if(_self.tip){
        _self.tip.hide();
      }
    });
  },
  showTip: function(obj){
    var _self = this;
    var title = $(obj).attr('datatitles');
    var left = 0, right=0, top = 0, flag = 10, position = 'bottom', arrowleft = 0;
    var offset = $(obj).offset();
    left = offset.left - $(obj).outerWidth(true) /2;
    top = offset.top + $(obj).outerHeight(true) + flag;
    arrowleft = $(obj).outerWidth(true);

    if(_self.tip){
      _self.tip.show({
        placement : position,
        arrowleft : arrowleft,
        title : title,
        left : left + 'px',
        right : right + 'px',
        top : top + 'px'
      });
      return;
    }
    if('' !== title){
      var tip = Tip._create({
        left: left,
        top: top,
        right: right,
        placement: position,
        events : {
          'aftercomplete' : function(){
            $('.tip-body', tip.el).html(title);
            var tipWidth = tip.el.outerWidth(true);
            if(arrowleft < tipWidth){
              var left = (arrowleft / tipWidth) * 100;
              $('.arrow', tip.el).css({left: left + '%'});
            }
            _self.emit('tiploaded', tip);
          }
        }
      });
      _self.tip = tip;
    }
  },
  selectNav: function(pathname, hostname, item){
    var _self = this;
    var sel = _self._selNav(pathname, hostname);
    var href = $(item).attr('href');
    var name = $(item).attr('name');
    if(href && href.indexOf(sel) > -1){
      $(item).addClass('btnsel');
      $(item).addClass(name+'sel');
      $(item).parent().addClass('selected');
    }
  },
  updateBindApp: function(){
    var _self = this;
    $('.toolbaricon', this.navbody).each(function(index, item){
      //替换url{appId}
      _self.beautifyUrl(item);
    });
  },
  beautifyUrl: function(obj){
    var name = obj.name;
    var url = $(obj).attr('dataurl') || '';
    var appId = localStorage.getItem('appId') || '';
    if(url.indexOf('{appId}') > -1){
      url = url.replace('{appId}', appId);
      if(appId === '' && name === 'ide'){
        url = url.replace(/\?.*/i, '');
      }
    }
    $(obj).attr('href', url);
  },
  _selNav: function(pathname, hostname){
    var sel = pathname;
    if(sel === '/s/cwf'){
      sel = '/s/cwf/prod';
    }else if(sel.indexOf('/dqc') > -1){
      sel = '/dqc';
    }else if(sel.indexOf('/webcollect') > -1){
      sel = '/webcollect';
    }else if(hostname.indexOf('ide') > -1){
      sel = 'ide';
    }
    return sel;
  },
  initSelNav : function(){
    var _self = this;
    var hostname = window.location.hostname;
    var sel = _self._selNav(window.location.pathname, hostname);
    var curNav = $('.item .btnsel', this.navbody);
    if(curNav.length === 0){
      //创建当前选中工具
      _self._getTools(function(err, tools){
        if(!err && tools){
          for(var i=0, len = tools.length; i < len; i ++){
            var itemTool = tools[i];
            var childrens = itemTool.children;
            if(childrens){
              for(var j=0, slen = childrens.length; j < slen; j ++){
                var sb = childrens[j];
                if(sb.path && sb.path.indexOf(sel) > -1){
                  var itemDom = $("<div class='item'>");
                  itemDom.appendTo(_self.navbody).insertBefore(_self.settingContainer);
                  var btnContainer = $("<div class='btnContainer selected'></div>").appendTo(itemDom);
                  var newA = $("<a></a>").appendTo(btnContainer);
                  newA.attr('name', sb.alias);
                  newA.attr('class', 'toolbaricon common-btn btnsel ticon-' + sb.alias + ' ' + sb.alias + 'sel');
                  newA.attr('datatitles', sb.name);
                  newA.attr('href', sb.path);
                  newA.attr('dataurl', sb.path);
                  newA.attr('target', '_blank');
                  _self.navBind(newA);
                  break;
                }
              }
            }
          }
        }
      });
    }
  }
});

exports = module.exports = Index;
