// webhook/.cov/run/sqlscan/res/page/selectresource/selectresource.js
var WindowPanel = require('hc-common/panel/windowpanel');
var Container = require('hc-common/panel/container');
var Store = require('hc-common/core/store');
var context = require('../../temp/context');
var Msg = require('hc-common/msg/msg');

var SelectreSource = function(){};
WindowPanel._extend(SelectreSource, {
  _className : 'selectresource',
  width : '800px',
  height : '550px',
  record : null,
  afterRender : function(){
    SelectreSource.prototype._super.afterRender.call(this);
    this._container = WinContainer._create({
      renderTo : this.body,
      win : this
    });
  }
});
var WinContainer = function(){};
Container._extend(WinContainer, {
  _className : 'selectresourcewin',
  record : null,
  _loadTpls : function(done){
    var _self = this;
    root.loadTpl('/page/selectresource/', 'selectresource', function(err, tpls){
      _self.tpl = tpls.selectresource;
      done();
    });
  },
  _loadCss : function(done){
    this._cssprefix = '/page/selectresource/';
    root.loadCss(this._cssprefix, ['selectresource'], function(){
      done();
    });
  },
  initData : function(){
    /*this.data = this.attr;
    this.msg = this.errMsg;*/
  },
  initEvents : function(){
    var _self = this;
    var btnSearch = this.query('search[name="btnSearch"]');
    btnSearch.on('click', function(){
      _self.loadSourceList($.trim(this.val()));
    });
    var btnSelectSourceSure = this.query('button[name="btnSelectSourceSure"]');
    btnSelectSourceSure.on('click', function(){
      _self.setResource();
    });
    var btnSelectSourceCancel = this.query('button[name="btnSelectSourceCancel"]');
    btnSelectSourceCancel.on('click', function(){
      if(_self.win){
        _self.win.doClose();
      }
      _self.win = null;
    });
  },
  afterComplete : function(){
    var _self = this;
    WinContainer.prototype._super.afterComplete.call(this);
    this.loadSourceList();
    var btnCreate = this.query('button[name=btnCreateSourceBySelect]');
    btnCreate.on('click', function(){
      _self.createResource()
    });
  },
  createResource : function(){
    var _self = this;
    if(this.win){
      this.win.doClose();
    }
    root.loadJs('/page/createresource/createresource.js', function(err, objs){
      _self.createWin = objs['/page/createresource/createresource.js']._create({
        title : '创建资源'
      });
      _self.emit('showcreateresourcewin');
    });
  },
  loadSourceList : function(resourcename){
    var _self = this;
    var sourcelist = this.query('list[name="sourcelist"]');
    var resourceName = resourcename || '';
    var opt = {
      url : context.urls.getResources,
      param : {
        resourceName : resourceName,
        author : '',
        resourceType : ''
      },
      callback : function(){
        _self.emit('afterresourcelistload');
      }
    };
    sourcelist.store.loadPage(1, opt);
  },
  setResource : function(record){
    var record = record || this.query('list[name="sourcelist"]').getSelectRecord();
    if(record.length === 0){
      Msg.alert('请选择资源名!');
      this.checkValueTest = '请选择资源名!';
      return;
    }
    var resourceLabel = $('#sqlScanResourceId');
    resourceLabel.attr("resourceId", record[0].id)
    resourceLabel.attr("value", record[0].name);
    this.checkValueTest = null;
    if(this.win){
      this.win.doClose();
    }
  }
});
exports = module.exports = SelectreSource;
