// webhook/.cov/run/sqlscan/res/temp/aliyun.js
/*!
 * aliyun 获取aliyun中的相关用户信息
 * dxpweb - lib/aliyun.js
 * Copyright(c) 2013 Taobao.com
 * Author: jifeng.zjd <jifeng.zjd@taobao.com>
 */

require('coffee-script');
var Client = require('easymysql');
var util = require('./util');
var ep = require('event-pipe');


var ErrorHand = function (err, sql, params, cb) {
  err.message += '\nsql:' + sql + '\nparams' + JSON.stringify(params);
  cb(err);
};

var Aliyun = function (options) {
  var servers = options.servers || [];
  var mysql = Client.create({
    'maxconnections' : 10 //最大连接数
  });
  for (var i = 0; i < servers.length; i++) {
    mysql.addserver(servers[i]);
  }
  this.mysql = mysql;
};

var INSERT_FLAG_SQL = 'INSERT INTO aliyun_user_flag(gmt_create, gmt_modified, \
  request_id) VALUES(now(), now(), :requestID);';
var SELECT_FLAG_SQL = 'SELECT id FROM aliyun_user_flag WHERE request_id = :requestID';
var DELETE_FLAG_SQL = 'DELETE FROM aliyun_user_flag WHERE request_id = :requestID'

Aliyun.prototype._getRandUserID = function (cb) {
  var mysql = this.mysql;
  var requestID = util.randRequestId();
  ep(function () {
    mysql.query({sql: INSERT_FLAG_SQL, params: {requestID: requestID}}, this);
  }, function (err) {
    if (err) {
      return ErrorHand(err, INSERT_FLAG_SQL, {requestID: requestID}, cb);
    };
    mysql.query({sql: SELECT_FLAG_SQL, params: {requestID: requestID}}, this);
  }, function (err, row) {
    if (err) {
      return ErrorHand(err, SELECT_FLAG_SQL, {requestID: requestID}, cb);
    }
    var id = row && row[0] && row[0].id;
    cb(null, id);
    mysql.query({sql: DELETE_FLAG_SQL, params: {requestID: requestID}}, function () {});
  }
  ).run();
};

var SELECT_USER_SQL = 'SELECT accessid, accesskey, displayname FROM aliyun_user WHERE id = :id AND status = 1';
var UPDATE_USER_SQL = 'UPDATE aliyun_user SET status = 0, workid = :workid, gmt_modified = now() WHERE id = :id;'

Aliyun.prototype._getRandUserByID = function (id, workid, cb) {
  var mysql = this.mysql;
  var data = null;
  ep(function () {
    mysql.query({sql: SELECT_USER_SQL, params: {id: id}}, this);
  }, function (err, row) {
    if (err) {
      return ErrorHand(err, SELECT_USER_SQL, {id: id}, cb);
    }
    data = row && row[0];
    mysql.query({sql: UPDATE_USER_SQL, params: {id: id, workid: workid}}, this);
  }, function (err) {
    if (err) {
     return ErrorHand(err, UPDATE_USER_SQL, {id: id, workid: workid}, cb);
    }
    return cb(null, data);
  }).run();
}

Aliyun.prototype.randUser = function (workid, cb) {
  var mysql = this.mysql;
  var self = this;
  ep(function () {
    self._getRandUserID(this);
  }, function (err, id) {
    if (err) {
      return cb(err);
    }
    self._getRandUserByID(id, workid, function (err, data) {
      if (err) {
        return cb(err);
      }
      if (!data) {
        return cb(new Error('data is null'));
      }
      cb(null, {
        odpsDisplayName: data.displayname,
        odpsAccessId: data.accessid,
        odpsAccessKey: data.accesskey
      });
    });
  }).run();
};

var COUNT_VALID_USER_SQL = 'SELECT count(*) FROM aliyun_user WHERE status = 1';
Aliyun.prototype.count = function (cb) {
  var mysql = this.mysql;
  mysql.query(COUNT_VALID_USER_SQL, function (err, rows) {
    if (err) {
      return cb(err);
    }
    cb(null, rows && rows[0] && rows[0]['count(*)']);
  });
}

module.exports = function (options) {
  return new Aliyun(options);
};
