// webhook/run/sqlscan/res/page/authorizerange/authorizerange.js
var WindowPanel = require('hc-common/panel/windowpanel');
var Container = require('hc-common/panel/container');
var Store = require('hc-common/core/store');
var context = require('../../temp/context');

var AuthorizeRange = function(){};
WindowPanel._extend(AuthorizeRange, {
  _className : 'authorizerange',
  width : '600px',
  height : '400px',
  record : null,
  afterRender : function(){
    AuthorizeRange.prototype._super.afterRender.call(this);
    this._container = WinContainer._create({
      renderTo : this.body,
      win : this,
      record : this.record
    });
  }
});
var WinContainer = function(){};
Container._extend(WinContainer, {
  _className : 'selectresourcewin',
  record : null,
  _loadTpls : function(done){
    var _self = this;
    root.loadTpl('/page/authorizerange/', 'authorizerange', function(err, tpls){
      _self.tpl = tpls.authorizerange;
      done();
    });
  },
  _loadCss : function(done){
    this._cssprefix = '/page/authorizerange/';
    root.loadCss(this._cssprefix, ['authorizerange'], function(){
      done();
    });
  },
  initData : function(){
    /*this.data = this.attr;
    this.msg = this.errMsg;*/
  },
  initEvents : function(){
    var _self = this;
    var projectList = _self.query('list[name="projectlist"]');
    projectList.on('rowclick', function(record){
      _self.loadUserInfoByProject(record, _self.record);
    });
    var btnSure = this.query('button[name="btnSure"]');
    btnSure.on('click', function(){
      _self.win.doClose();
    });
  },
  afterComplete : function(){
    var _self = this;
    WinContainer.prototype._super.afterComplete.call(this);
    this.loadProjectList();
  },
  loadProjectList : function(){
    var _self = this;
    var projectList = _self.query('list[name="projectlist"]');
    projectList.store.load({
      url : context.urls.getProNameByRuleId,
      param : {
        id : _self.record.id,
        status : _self.record.status
      },
      callback : function(err, rows){
        projectList.selectRowByIndex(0);
        _self.emit('afterprojectlistload');
      }
    });
  },
  loadUserInfoByProject : function(record, ruleRecord){
    var _self = this;
    var userList = this.query('list[name="userlist"]');
    userList.store.load({
      url : context.urls.getUserNameByProId,
      param : {
        projectName : record.projectName,
        ruleId : ruleRecord.id,
        status : ruleRecord.status
      },
      callback : function(err, rows){
        _self.emit('afteruserinfolistload');
      }
    });
  }
});
exports = module.exports = AuthorizeRange;
