// webhook/.cov/run/sqlscan/res/page/remotepathdev.js
exports = module.exports = function(opt) {
  return {
    fnMenu : '/mk/cheetah/fnmenu.js',
    //责任人
    owners : 'mk/metadata/getUsers.json',
    //获取所有应用
    apps : '/mk/apps.json',
    //获取我的应用
    myapps : '/mk/getApplications.json',
    //获取我的应用，包含role信息
    myappInfos : '/mk/getApplications.json',
    //根据应用ID获取当前应用下的人
    appUsers : '/mk/appusers.json',
    //根据工号ID获取应用列表
    getApplicationByUser: '/mk/cheetah/getApplicationByUser.json',
    //获取list列表-查看规则
    getRuleListView : '/mk/getRules.json',
    //获取list列表-管理规则
    getRuleListAdd : '/mk/getRules.json',
    //获取规则名列表
    rulenames : '/mk/rulenames.json',
    //根据规则ID获取规则信息
    getRuleInfoById : '/mk/getRuleInfoById.json',
    //新增规则
    insertRuleAndConfigure : '/mk/insertRuleAndConfigure.json',
    //编辑规则
    updateRule : '/mk/updateRuleAndDeleteRule.json',
    //删除规则
    deleteRule : '/mk/updateRuleAndDeleteRule.json',
    //恢复规则
    recoverRule : '/mk/updateRuleAndDeleteRule.json',
    //根据规则ID获取对应的授权应用
    getProNameByRuleId : '/mk/getProNameByRuleId.json',
    //根据project ID获取对应的授权用户
    getUserNameByProId : '/mk/getUserNameByProId.json',
    //获取资源信息
    getResources : '/mk/getResources.json',
    //上传资源到tfs
    uploadResourceTotfs : '/mk/uploadTfs.json',
    //添加资源
    insertResource : '/mk/insertResource.json',
    //获取projectlist
    getProjects : '/mk/getProjects.json',
    //根据projectId启用project的规则
    enableProjectRuleByProjectId : '/mk/updateRuleAndDeleteRule.json',
    //获取projectlist
    disableProjectRuleByProjectId : '/mk/updateRuleAndDeleteRule.json'
  };
};
