// webhook/.cov/run/sqlscan/res/page/remotepath.js
exports = module.exports = function(opt) {
  var prex = '/r/';
  return {
    fnMenu : '/s/menu',
    //责任人
    //owners: prex + 'process/ProcessAPI/getAllUser',
    owners: prex + 'apps/user/UserAPI/getUsers',
    //获取所有应用
    apps: prex + 'authorization/ProjectManager/getPageModel',
    //获取我的应用-CUC
    myapps : prex + 'apps/application/ApplicationAPI/getApplicationByUser',
    //获取我的应用-sqlscan，包含role信息
    myappInfos : prex + 'sqlscan/CUCService/getApplicationByUid',
    //根据应用ID获取当前应用下的人
    appUsers : prex + 'apps/application/ApplicationAPI/getAppMemberInfo',
    //根据工号ID获取应用列表
    getApplicationByUser: prex + 'application/ApplicationAPI/getApplicationByUser',
    //获取list列表-查看规则
    getRuleListView : prex + 'sqlscan/RuleManageService/getRulesBy',
    //获取list列表-管理规则
    getRuleListAdd : prex + 'sqlscan/RuleManageService/getRules',
    //获取规则名列表
    rulenames : prex + 'sqlscan/RuleManageService/getRuleNames',
    //根据规则ID获取规则信息
    getRuleInfoById : prex + 'sqlscan/RuleManageService/getRulesById',
    //新增规则
    insertRuleAndConfigure : prex + 'sqlscan/RuleManageService/insertRuleAndConfigure',
    //编辑规则
    updateRule : prex + 'sqlscan/RuleManageService/updateRule',
    //删除规则
    deleteRule : prex + 'sqlscan/RuleManageService/deleteRule',
    //恢复规则
    recoverRule : prex + 'sqlscan/RuleManageService/recoverRule',
    //根据规则ID获取对应的授权应用
    getProNameByRuleId : prex + 'sqlscan/ConfigureManageService/getProNameBy',
    //根据project ID获取对应的授权用户
    getUserNameByProId : prex + 'sqlscan/ConfigureManageService/getListUserBy',
    //获取资源信息
    getResources : prex + 'sqlscan/ResourceManageService/getResources',
    //上传资源到tfs
    uploadResourceTotfs : prex + 'sqlscans/upload',
    //添加资源
    insertResource : prex + 'sqlscan/ResourceManageService/insertResource',
    //获取projectlist
    getProjects : prex + 'sqlscan/CUCService/getApplicationByUidAppId',
    //根据projectId启用project的规则
    enableProjectRuleByProjectId : prex + 'sqlscan/CUCService/openSqlscan',
    //获取projectlist
    disableProjectRuleByProjectId : prex + 'sqlscan/CUCService/closeSqlscan'
  };
};
