// webhook/.cov/run/sqlscan/res/page/rulemanage/rulemanage.js
var Container = require('hc-common/container/container');
var context = require('../../temp/context');
var eventProxy = require('hc-common/core/eventproxy');
var Store = require('hc-common/core/store');
var Msg = require('hc-common/msg/msg');

var Rulemanage = function(){};

Container._extend(Rulemanage, {
  _className : 'rulemanage',
  __filename: __filename,
  isHide : false,
  initData : function(){
    this.data = {
      url : context.urls
    };
  },
  initEvents : function(){
    this.searchEvent();
    this.rowClickEvent();
    this.tabEvent();
  },
  afterComplete : function(){
    Rulemanage.prototype._super.afterComplete.call(this);
    var owner = this.query('combobox[name=owner]');
    owner.query = function(searchText){
      return {
        query : JSON.stringify({
          keywords : $.trim(searchText)
        })
      };
    };
    var ruleName = this.query('combobox[name=rulename]');
    ruleName.query = function(searchText){
      return {
        query : searchText
      };
    };
    this.loadRankStore();
    this.loadRuleManageList();
  },
  searchEvent : function(){
    var _self = this;
    var btnSearch = _self.query('button[name="btnSearch"]');
    btnSearch.on('click', function(){
      _self.loadRuleManageList();
    });
    var btnAddRule = _self.query('button[name="btnAddRule"]');
    btnAddRule.on('click', function(){
      eventProxy.emit('hash.switch', {
        controller : 'ruleadd'
      });
    });
  },
  rowClickEvent : function(){
    var _self = this;
    var rulemanagelist = _self.query('list[name="rulemanagelist"]');
    rulemanagelist.on('rowclick', function(record, row, e){
      if($(e.target).attr('class') === 'rule-desc'){
        _self.loadRuleDescDetail(record, $(e.target), this);
      }else if($(e.target).attr('value') === "authorization"){
        _self.viewAuthorizeRange(record);
      }else if($(e.target).attr('class') === 'btnUpdate'){
        _self.getRuleInfoById(record.id, function(ruleInfo){
          ruleInfo = JSON.stringify(ruleInfo);
          _self.checkUpdateRule = true;
          _self.emit('beginupdaterule');
          eventProxy.emit('hash.switch', {
            controller : 'ruleadd',
            ruleInfo : ruleInfo
          });
        });
      }else if($(e.target).attr('class') === 'btnDelete' || $(e.target).attr('class') === 'btnRestore'){
        _self.deleteOrRestoreRule(record);
      }
    })
  },
  tabEvent : function(){
    var _self = this;
    var tab = $('.rulemanagelist-tab', this.el).children("span");
    tab.on('click' ,function(){
      if($(this).attr('class') === 'rulemanagelist-tab-active'){
        return;
      }
      for(var i = 0, len = tab.length; i < len; i++){
        if(tab[i] === this){
          tab[i].className = "rulemanagelist-tab-active"
        }else{
          tab[i].className = "rulemanagelist-tab-deactive"
        }
      }
      _self.loadRuleManageList();
    });
  },
  loadRankStore : function(){
    var grade = this.query('combobox[name=rank]');
    var data = [{
      value : '',
      text : '全部'
    },{
      value : 1,
      text : '提醒'
    }, {
      value : 2,
      text : '禁止'
    }];
    grade.store.load(data);
  },
  loadRuleManageList : function(){
    var _self = this;
    var rulemanagelist = _self.query('list[name="rulemanagelist"]');
    var ruleName = _self.query('combobox[name="rulename"]').val();
    var owner = _self.query('combobox[name="owner"]').val();
    var rank = _self.query('combobox[name="rank"]').val();
    var status = _self.getActiveStatus();
    var opt = {
      url : context.urls.getRuleListAdd,
      param : {
        ruleName : ruleName,
        owner : owner,
        rank : rank,
        status : status
      },
      callback : function(){
        var count = rulemanagelist.store.recordCount;
        var label = $('#rulecount');
        label.html('该条件下共有规则 ' + count +' 条');
        _self.emit('afterrulemanagelistload');
      }
    };
    var odpsDisplayName = context.options.session.user.odpsDisplayName;
    rulemanagelist.beforeRowRender = function(record){
      if(odpsDisplayName === record.authorUserName){
        record.isOwner = true;
      }
    }
    rulemanagelist.store.loadPage(1, opt);
  },
  loadRuleDescDetail : function(record, desc, row){
    if(record.detail.length <= 50){
      return;
    }
    var detail = $(desc).parent().find(".rule-detail");
    if(!row.showDesc){
      //detail.html(record.activityDefinitionId);
      detail.slideDown();
      row.showDesc = true;
      this.checkRowClick = true;
      return;
    }
    detail.slideUp();
    this.checkRowClick = null;
    row.showDesc = null;
  },
  viewAuthorizeRange : function(record){
    var _self = this;
    root.loadJs('/page/authorizerange/authorizerange.js', function(err, objs){
      _self.win = objs['/page/authorizerange/authorizerange.js']._create({
        title : '授权范围',
        record : record
      });
      _self.emit('showauthorizerange');
    });
  },
  deleteOrRestoreRule : function(record){
    var _self = this, userInfo = context.options.session.user;
    var url = context.urls.deleteRule;
    var index = 1;
    var message = '资源删除成功';
    if(record.status === 1){
      url = context.urls.recoverRule;
      index = 0;
      message = "资源恢复成功"
    }
    var ruleIds = [];
    ruleIds.push(JSON.stringify({
      "long" : record.id.toString()
    }));
    var store = Store._create({
      url : url,
      param : {
        ruleIds : ruleIds,
        odpsKey : userInfo.odpsDisplayName,
      }
    });
    store.load(function(err, data){
      if(err){
        return;
      }
      _self.setActiveStatus(index);
      _self.loadRuleManageList();
      _self.emit('afterdeleterule');
      Msg.alert(message);
    });
  },
  getActiveStatus : function(){
    var tab = $('.rulemanagelist-tab');
    var status = tab.find('.rulemanagelist-tab-active').attr('value');
    return status;
  },
  setActiveStatus : function(index){
    var tab = $('.rulemanagelist-tab').children("span");
    for(var i = 0, len = tab.length; i<len ; i++){
      if(i === index){
        tab[i].className = 'rulemanagelist-tab-active';
      }else{
        tab[i].className = 'rulemanagelist-tab-deactive';
      }
    }
  },
  getRuleInfoById : function(ruleId, cb){
    var store = Store._create({
      url : context.urls.getRuleInfoById,
      param : {
        ruleId : ruleId,
      }
    });
    store.load(function(err, data){
      if(err){
        return;
      }
      cb(data);
    });
  }
});

exports = module.exports = Rulemanage;
