// webhook/.cov/run/sqlscan/res/page/common/userinfo/index.js
var WindowPanel = require('../../../com/window/windowpanel');
var context = require('../context');
var Store = require('../../../core/store');
var $ = require('jquery');

var Index = function(){};
WindowPanel._extend(Index, {
  _className : 'userInfo',
  tplName: 'page/common/userinfo/index',
  cssName: 'page/common/userinfo/index',
  prefix: '',
  afterComplete : function(){
    Index.prototype._super.afterComplete.call(this);
    this.getUserOdpsInfo();
  },
  getUserOdpsInfo : function(){
    var _self = this;
    if(_self.prefix === ''){
      var store = new Store();
      store.load({
        url : context.urls.getUserOdpsInfo,
        callback : function(err, userInfo){
          if(!err){
            $('.odpsAccessKeyInfo .odpsDisplayName', _self.el).html(userInfo.odpsDisplayName);
            $('.odpsAccessKeyInfo .odpsAccessId', _self.el).html(userInfo.odpsAccessId);
            $('.odpsAccessKeyInfo .odpsAccessKey', _self.el).html(userInfo.odpsAccessKey);
          }
        }
      });
    }else{
      _self._getHeaderInfo(_self.prefix, function(err, obj){
        if(!err){
          var userInfo = obj.userOpdsInfo;
          $('.odpsAccessKeyInfo .domainUser', _self.el).html(obj.user.DomainUser);
          $('.odpsAccessKeyInfo .odpsDisplayName', _self.el).html(userInfo.odpsDisplayName);
          $('.odpsAccessKeyInfo .odpsAccessId', _self.el).html(userInfo.odpsAccessId);
          $('.odpsAccessKeyInfo .odpsAccessKey', _self.el).html(userInfo.odpsAccessKey);
        }
      });
    }
  },
  _getHeaderInfo: function(prefix, cb){
    var url = prefix + '/nr/header/view';
    $.ajax({
      type: 'GET',
      url: url,
      dataType:'jsonp',
      async: false,
      success: function(data){
        if(data){
          data = data.replace(/\n|\r/g, '');
          data = data.replace(/\\/g, '\\\\');
          var obj = JSON.parse(data);
          cb(null, obj);
        }
      },
      error: function(jqXHR, textStatus, errorThrown){
        cb(errorThrown);
      }
    });
  }
});
exports = module.exports = Index;
