// webhook/.cov/run/sqlscan/res/page/common/toolbar/index.js
var $ = require('jquery');
var Container = require('../../../com/container/container');
var eventProxy = require('../../../core/eventproxy');
var context = require('../context');
var Helper = require('../helper/index');

var Index = function(){};
Container._extend(Index, {
  _className : 'index',
  __filename: __filename,
  prefix: '',
  atatech: '',
  logoIsLeft: false,
  isHelperHide: false,
  _loadStore: function(done){
    var _self = this;
    var prefix = _self._getPrefix();
    var appId = _self.getUrlParam(document.domain, window.location.href);
    this._getHeaderInfo(prefix, appId, function(err, data){
      if(!err){
        _self.headerInfo = data;
        localStorage.setItem('appId', data.lastApp.lastAppId);
        done();
      }
    });
  },
  afterRender: function(){
    Index.prototype._super.afterRender.call(this);
    this.mininav = $('.mininav', this.el);
    if(this.logoIsLeft){
      $('.container', this.mininav).css({width: '100%'});
      $('.container .logo', this.mininav).css({'margin-left': '22px'});
      $('.container .loginAll', this.mininav).css({'margin-right': '22px','width':'auto'});
    }
  },
  afterComplete: function(){
    Index.prototype._super.afterComplete.call(this);
    var _self = this;
    this.btnDoc = this.query("*[name=doc]");
    this.btnMsg = this.query("*[name=message]");
    this.btnStting = this.query("*[name=setting]");
    this.navConiner = this.query("*[name=navs]");
    this.myappContain = this.query('combobox[name=myApps]');
    this.mininav = $('.mininav', this.el);
    this.navs = $('.navs', this.mininav);
    this.logo = $('.logo', this.el);
    this.offsetTop = 10;
    var prefix = _self._getPrefix();
    if(this.headerInfo){
      var data = _self.headerInfo;
      _self.loadMyApps(data.myApps, data.lastApp.lastAppId);
      var hash = localStorage.getItem('dxp_toolbar_notices');
      var notices = data.notices;
      if(notices.hash !== hash){
        _self.btnMsg.el.find('.btn').addClass('notify');
      }
      _self.cacheMyToolIds(data.myTools);
    }
    _self.filterApp();
    eventProxy.on('trigger.update.app', function(appId){
      _self.triggerUpdateApp(true);
    });
    var navs = this.query("*[name=navs]");
    _self.on('cwf.env.change', function(env){
      navs.emit('cwf.env.change', env);
    });
    if(this.isHelperHide){
      var options = {
        user: _self.headerInfo.user,
        prefix: _self._getPrefix()
      };
      new Helper(options);
    }
    //更新最近操作应用Id
    // _self.updateLastApp();
  },
  cacheMyToolIds: function(objs){
    var ids = [], item;
    if(objs){
      for(var i=0, len = objs.length; i < len; i ++ ){
        item = objs[i];
        ids.push(item.id);
      }
    }
    sessionStorage.setItem('global_toolbar_mytool_ids', ids.join(','));
  },
  initData: function(){
    var prefix = this._getPrefix();
    var _self = this;
    this.data ={
      prefix : prefix,
      navFlag: _self.logoIsLeft,
      nick : _self.headerInfo.user.nickNameCn || _self.headerInfo.user.nick || _self.headerInfo.user.WorkId
    };
  },
  _getPrefix: function(){
    var prefix = '';
    var _self = this;
    if(context && context.options){
      if(context.options.prefix){
        prefix = context.options.prefix;
      }else{
        prefix = _self.prefix;
      }
    }else if(this.prefix){
      prefix = _self.prefix;
    }else{
      prefix = '';
    }
    return prefix;
  },
  initEvents: function(){
    Index.prototype._super.initEvents.call(this);
    var _self = this;
    this.docEvent();
    this.notice();
    this.settingEvent();
    if(this.logo !== undefined){
      this.logo.bind('click', function(){
        window.location.href = _self._getPrefix() +'/s/cheetah';
      });
    }
    this.query('combobox[name=myApps]').on('change', function(){
      _self._selectApp();
    });
  },
  docEvent: function(){
    var opt = null, offset = null, _self = this, btnWidth, bodyWidth, left, docNav = null, popoverBody = null, tcontiner, minLeft;
    offset = this.btnDoc.el.offset();
    var pluginfn = function(obj){
      var propover = require('../../../newcom/popover/index.js')._create({
        events: {
          'aftercomplete': function(){
            popoverBody = $('.popover-content', propover.el);
            docNav = require('../docnav/index.js')._create({
              renderTo: popoverBody,
              prefix: obj.prefix,
              atatech : obj.atatech,
              events: {
                'aftercomplete' : function(){
                  tcontiner = $('.mininav .container');
                  if(tcontiner.offset()){
                    left = tcontiner.offset().left + _self.btnMsg.el.outerWidth(true) + _self.btnStting.el.outerWidth(true);
                  }
                  if(_self.logoIsLeft){
                    left += 22;
                  }
                  bodyWidth = $(document.body).outerWidth(true);
                  right = left;
                  btnWidth = _self.btnDoc.el.width();
                  opt = {
                    top : _self.offsetTop + offset.top + _self.btnDoc.el.outerHeight(true),
                    right: right
                  };
                  propover.show(opt);
                  popoverWidth = popoverBody.width();
                  minLeft = bodyWidth - popoverWidth - right;
                  propover.setArrowPosition(offset.left - minLeft + btnWidth/3);
                  propover.emit('docnav.loaded');
                }
              }
            });
          }
        }
      });
      return propover;
    };
    this.btnDoc.on('click', function(){
      var btn = _self.btnDoc.el.find('.btn');
      var _prefix = _self._getPrefix();
      var atatech = context.options.atatech || _self.atatech;
      if(btn.hasClass('selected')){
        _self.propoverDestroy();
      }else{
        _self.propoverDropdown();
        var obj = {
          prefix: _prefix,
          atatech : atatech
        };
        _self.propover = pluginfn(obj);
        $(window).bind('click.dropup', function(e){
          if(btn.hasClass('selected') && !$.contains(_self.el[0], e.target)){// !_self.el.find(e.target).length){
            _self.propoverDestroy();
            _self.btnDoc.el.find('.doc').trigger('click');
          }
        });
      }
    });
    this.btnDoc.el.find("button.js-dropdown").on('dropup', function(){
      _self.propoverDestroy();
      _self.btnDoc.el.find('.doc').removeClass('selected');
    });
  },
  notice: function(){
    var opt = null, offset = null, _self = this,
    notify = null, popoverBody = null, flag = false,tcontiner, left, right, bodyWidth, minLeft, popoverWidth, btnWidth;
    var pluginfn = function(){
      var popoverWidth;
      var propover = require('../../../newcom/popover/index.js')._create({
        events: {
          'aftercomplete': function(){
            popoverBody = $('.popover-content', propover.el);
            notify = require('../notify/index.js')._create({
              renderTo: popoverBody,
              events: {
                'aftercomplete': function(){
                  var prefix = _self._getPrefix();
                  if(_self.headerInfo){
                    eventProxy.emit('toolbar.notices', null, _self.headerInfo.notices);
                  }
                  tcontiner = $('.mininav .container');
                  if(tcontiner.offset()){
                    left = tcontiner.offset().left;
                  }
                  if(_self.logoIsLeft){
                    left += 22;
                  }
                  bodyWidth = $(document.body).outerWidth(true);
                  right = left;
                  offset = _self.btnMsg.el.offset();
                  btnWidth = _self.btnMsg.el.width();
                  opt = {
                    top : _self.offsetTop + offset.top + _self.btnMsg.el.outerHeight(true),
                    right: right
                  };
                  propover.show(opt);
                  popoverWidth = popoverBody.width();
                  minLeft = bodyWidth - popoverWidth - right;
                  propover.setArrowPosition(offset.left - minLeft + btnWidth/3);
                  propover.emit('notify.loaded');
                }
              }
            });
          }
        }
      });
      return propover;
    };
    this.btnMsg.on('click', function(){
      var btn = _self.btnMsg.el.find('.btn');
      btn.removeClass('notify');
      if(btn.hasClass('selected')){
        _self.propoverDestroy();
      }else{
        _self.propoverDropdown();
        _self.propover = pluginfn();
        $(window).bind('click.dropup', function(e){
          if(btn.hasClass('selected') && !$.contains(_self.el[0], e.target)) { //!_self.el.find(e.target).length){
            _self.propoverDestroy();
            _self.btnMsg.el.find('.mail').trigger('click');
          }
        });
      }
    });
    this.btnMsg.el.find("button.js-dropdown").on('dropup', function(){
      _self.propoverDestroy();
      _self.btnMsg.el.find('.mail').removeClass('selected');
    });
  },
  settingEvent: function(){
    var opt = null, offset = null, _self = this, settingdom = null,
    popoverBody = null, flag = false, tcontiner, left, bodyWidth, right, minLeft;
    var pluginfn = function(prefix){
      var popoverWidth;
      var propover = require('../../../newcom/popover/index.js')._create({
        events: {
          'aftercomplete': function(){
            popoverBody = $('.popover-content', propover.el);
            settingdom = require('../settings/index.js')._create({
              renderTo: popoverBody,
              prefix: prefix,
              events: {
                'aftercomplete' : function(){
                  tcontiner = $('.mininav .container');
                  if(tcontiner.offset()){
                    left = tcontiner.offset().left;
                  }
                  if(_self.logoIsLeft){
                    left += 22;
                  }
                  bodyWidth = $(document.body).outerWidth(true);
                  right = left;
                  offset = _self.btnStting.el.offset();
                  opt = {
                    top : _self.offsetTop + offset.top + _self.btnStting.el.outerHeight(true),
                    right: right
                  };
                  var btnWidth = _self.btnStting.el.width();
                  propover.show(opt);
                  popoverWidth = popoverBody.width();
                  minLeft = bodyWidth - popoverWidth - right;
                  propover.setArrowPosition(offset.left - minLeft + btnWidth/2);
                  propover.emit('settingdom.loaded');
                }
              }
            });
          }
        }
      });
      return propover;
    };
    this.btnStting.on('click', function(){
      var btn = _self.btnStting.el.find('.btn');
      var _prefix = _self._getPrefix();
      if(btn.hasClass('selected')){
        _self.propoverDestroy();
      }else{
        _self.propoverDropdown();
        _self.propover = pluginfn(_prefix);
        $(window).bind('click.dropup', function(e){
          if(btn.hasClass('selected') && !$.contains(_self.el[0], e.target)) { //!_self.el.find(e.target).length){
            _self.propoverDestroy();
            _self.btnStting.el.find('.setting').trigger('click');
          }
        });
      }
    });
    this.btnStting.el.find("button.js-dropdown").on('dropup', function(){
      _self.propoverDestroy();
      _self.btnStting.el.find('.setting').removeClass('selected');
    });
  },
  _getHeaderInfo: function(prefix, appId, cb){
    var url = prefix + '/nr/header/view';
    if(appId){
      url += '?appId=' + appId;
    }
    $.ajax({
      type: 'GET',
      url: url,
      dataType:'jsonp',
      async: false,
      success: function(data){
        if(data){
          data = data.replace(/\n|\r/g, '');
          data = data.replace(/\\/g, '\\\\');
          var obj = JSON.parse(data);
          cb(null, obj);
        }
      },
      error: function(jqXHR, textStatus, errorThrown){
        cb(errorThrown);
      }
    });
  },
  loadMyApps : function(record, app){
    var myApps = this.query('combobox[name=myApps]');
    myApps.store.load(record);
    if(app){
      myApps.val(app);
    }else{
      if(myApps.records.length > 0){
        myApps.valByIndex(0);
        localStorage.setItem('appId', myApps.val());
        this.navConiner.updateBindApp();
      }
    }
  },
  propoverDropdown: function(){
    $('button.js-dropdown').trigger('dropup');
  },
  propoverDestroy: function(){
    var _self = this;
    if(_self.propover){
      _self.propover.destroy();
      _self.propover = null;
    }
    $(window).unbind('click.dropup');
  },
  _selectApp: function(){
    var $myapp, appId, curNav, href, cls, _self = this;
    $myapp = this.query('combobox[name=myApps]');
    appId = $myapp.val();
    localStorage.setItem('appId', appId);
    curNav = $('button.btnsel', this.navs);
    cls = curNav.attr('class');
    if(!cls){
      cls = '';
    }
    var sel = window.location.pathname;
    var hostname = window.location.hostname;
    if(hostname.indexOf('ide') > -1 || sel.indexOf('app/task') > -1){
      this.emit('myapp.change', appId);
    }else if(cls.indexOf('ticon-appConfig') > -1 ||  sel.indexOf('cheetah/app') > -1){
      eventProxy.emit('menu.switch', {
        controller : 'manage',
        app : appId
      });
      _self.triggerUpdateApp(true);
    }
    _self.navConiner.updateBindApp();
  },
  triggerUpdateApp: function(flag, appId){
    var _self = this, myapp = this.query('combobox[name=myApps]');
    if(flag){
      if(appId){
        _self.updateLastApp(appId);
      }else{
        _self.updateLastApp();
      }
      _self.navConiner.updateBindApp();
    }else{
      localStorage.setItem('appId', appId);
      myapp.setVal(appId, null, false);
    }
  },
  filterApp: function(){
    var $myapp, cls, curNav;
    $myapp = this.query('combobox[name=myApps]');
    var notedit = $('.notedit', $myapp.el);
    curNav = $('button.btnsel', this.navs);
    cls = curNav.attr('class');
    if(!cls){
      cls = '';
    }
    var sel = window.location.pathname;
    var hostname = window.location.hostname;
    if(cls.indexOf('ticon-ide') > -1 || cls.indexOf('ticon-appConfig') > -1 || sel.indexOf('app/task') > -1 || sel.indexOf('cheetah/app') > -1 || hostname.indexOf('ide') > -1){
      $('.allApp', notedit).remove();
      $('.field-container', $myapp.el).show();
    }else{
      $('.field-container', $myapp.el).hide();
      notedit.append("<input type='text' value='全部应用' readonly='true' class='allApp'/>");
    }
  },
  updateLastApp: function(id){
    var _self = this, domain = document.domain, href = window.location.href;
    var appId = _self.getUrlParam(domain, href);
    if(id){
      appId = id;
    }
    _self._getHeaderInfo(_self._getPrefix(), appId, function(err, data){
      if(!err){
        _self.headerInfo = data;
        var myApps = _self.query('combobox[name=myApps]');
        myApps.setVal(data.lastApp.lastAppId, null, false);
        localStorage.setItem('appId', data.lastApp.lastAppId);
      }
    });
  },
  getUrlParam: function(domain, href){
    var regExp, appId ;
    var map = [
      {domain : /^dw-pre(-\w+)?\./i, match: /\?.*projectid=(\d+)/im},
      {domain : /\bide\d?\./i, match: /\?.*projectid=(\d+)/im},
      {domain : /^ndxp(-\w+)?\./i, match: /\#.*app:(\d+)/im},
    ];
    var domainRule, matchRule, res;
    for (var i = 0, len = map.length; i < len; i++ ){
      domainRule = map[i].domain;
      matchRule = map[i].match;
      if (domain.match(domainRule)) {
        if (res = href.match(matchRule)){
          return res[1];
        } else {
          return null;
        }
      }
    }
    return null;
  }
});
exports = module.exports = Index;
