// webhook/run/sqlscan/res/page/common/settings/index.js
var $ = require('jquery');
var Container = require('../../../com/container/container');
var context = require('../context');
var eventProxy = require('../../../core/eventproxy');

var Index = function(){};

Container._extend(Index, {
  _className : 'index',
  __filename: __filename,
  prefix: '',
  initData: function(){
    var _self = this;
    this.data = {
      prefix: _self.prefix
    };
  },
  afterComplete: function(){
    Index.prototype._super.afterComplete.call(this);
    var _self = this;
    this.userInfo = $('.userInfo', this.el);
  },
  initEvents: function(){
    Index.prototype._super.initEvents.call(this);
    var _self = this;
    this.userInfo.on('click', function(){
      _self.showUserInfo();
    });
  },
  showUserInfo : function(){
    var win, _self = this;
    root.loadJs('/module/page/common/userinfo/index.js', function(err, objs){
      win = objs['/module/page/common/userinfo/index.js']._create({
        title: '个人信息',
        width: '600px',
        height:'450px',
        prefix: _self.prefix
      });
    });
  }
});

exports = module.exports = Index;
