// webhook/run/sqlscan/res/page/common/pluginsetting/index.js
var $ = require('jquery');
var Container = require('../../../com/container/container');
var context = require('../context');
var Store = require('../../../core/store');
var eventProxy = require('../../../core/eventproxy');
var context = require('../context');

var Index = function(){};
Container._extend(Index, {
  _className : 'index',
  __filename: __filename,
  prefix: '',
  _loadStore: function(done){
    var _self = this;
    if(localStorage.getItem('global_toolbar_all_tools')){
      _self.pluginList = _self.parseString(localStorage.getItem('global_toolbar_all_tools'));
      done();
    }else{
      this._getTools(function(err, data){
        if(!err){
          _self.pluginList = data;
          done();
        }
      });
    }
  },
  afterComplete: function(){
    Index.prototype._super.afterComplete.call(this);
    var _self = this;
    this.checkboxArr = $('.checkbox', this.el);
    this.pluginIcon = $('.pluginicon', this.el);
    this.pluginDesc = $('.plugindesc', this.el);
    this.photo = $('.photo', this.el);
    this.btnSetting = this.query("*[name='btnSetting']");
    this.triggerToolsChange();
  },
  triggerToolsChange: function(){
    var _self = this;
    eventProxy.on('tools.change', function(ids){
      if(ids.length >= 6){
        _self.checkboxArr.each(function(index, element){
          var id = $(element).attr('value');
          if($(element).hasClass('dashboard')){
            $(element).addClass('checked');
          }
          if(ids.indexOf(id) === -1){
            $(element).addClass('disabled');
            $(element).removeClass('checked');
            if(_self.tip){
              _self.tip.destroy();
              _self.tip = null;
              _self._removeEvent();
            }
            _self._disabledEvent($(element), '只能设置6个常用工具哦~');
          }
        });
      }else{
        _self.checkboxArr.each(function(index, element){
          var id = $(element).attr('value');
          if(ids.length < 1){
            _self._disabledEvent($(element), '请至少选择一个常用工具~');
          }else{
            $(element).removeClass('disabled');
            if(_self.tip){
              _self.tip.destroy();
              _self.tip = null;
            }
            _self._removeEvent();
          }
        });
      }
    });
  },
  _removeEvent: function(){
    this.checkboxArr.each(function(index, element){
      $(element).unbind('mouseenter');
      $(element).unbind('mouseleave');
    });
  },
  _disabledEvent: function(element, tipContent){
    this.tip = null;
    var _self = this;
    $(element).on('mouseenter', function(){
      var offset = $(element).offset();
      $('.tip', $(document.body)).remove();
      _self.tip = require('../../../newcom/tip/index')._create({
        left: offset.left,
        top: offset.top + $(element).outerHeight(true),
        placement: 'bottom',
        hidearrow: 'false',
        events : {
          'aftercomplete' : function(){
            $('.tip-body', _self.tip.el).html(tipContent);
          }
        }
      });
    });
    $(element).on('mouseleave', function(){
      if(_self.tip){
        _self.tip.destroy();
        _self.tip = null;
      }
    });
  },
  initData: function(){
    var _self = this;
    _self.data = {
      pluginList : _self.pluginList
    };
    _self.emit('initdta.loaded');
  },
  initEvents: function(){
    Index.prototype._super.initEvents.call(this);
    var _self = this;
    _self._pluginHover();
    _self._settingTools();
    _self._pluginClick();
  },
  saveMyTools: function(cb){
    var _self = this;
    if(_self.clickCount && _self.clickCount.length > 0){
      var ids = _self.clickCount.join(',');
      _self._updateTools(ids, function(err, data){
        if(!err){
          cb(data.returnValue);
        }
      });
    }else{
      window.alert('请至少选择一个常用工具~');
      cb(false);
    }
  },
  _settingTools: function(){
    var _self = this;
    _self.settingFlag = false;
    this.btnSetting.on('click', function(){
      if(_self.settingFlag){
        _self.saveMyTools(function(flag){
          if(flag){
            eventProxy.emit('tools.destroy');
            eventProxy.emit('tools.bindclick');
          }
        });
      }else{
        _self.btnSetting.val('完成设置');
        var btn = $('.btn', _self.btnSetting.el);
        btn.removeClass('setting');
        btn.parent().addClass('selected');
        _self.settingFlag = true;
        _self.clickCount = [];

        var cacheIds = sessionStorage.getItem('global_toolbar_mytool_ids');
        if(cacheIds){
          _self.checkboxArr.each(function(index, element){
            var id = $(element).attr('value');
            var data = cacheIds.split(',');
            data = data.length > 6 ? data.slice(0,6) : data;
            if(data.indexOf(id) > -1){
              $(element).addClass('check');
              $(element).addClass('checked');
              _self.clickCount.push(id);
            }else{
              $(element).addClass('check');
            }
          });
          // eventProxy.emit('tools.change', _self.clickCount, _self._getSelTools(_self.clickCount));
        }
      }
    });
  },
  _pluginClick: function(){
    var _self = this;
    this.checkboxArr.each(function(index, element){
      //替换url{appId}
      _self.beautifyUrl(element);
      $(element).on('click', function(e){
        e.stopImmediatePropagation();
        var id = $(element).attr('value');
        if(_self.settingFlag ){
          var index = _self.clickCount.indexOf(id);
          if(_self.clickCount.length <= 5 ){
            if(index > -1){
              _self.clickCount = _self.removeArr(_self.clickCount, index);
            }else{
              _self.clickCount.push(id);
            }
          }else{
            if(index > -1){
              _self.clickCount = _self.removeArr(_self.clickCount, index);
            }
          }
          eventProxy.emit('tools.change', _self.clickCount, _self._getSelTools(_self.clickCount));
          sessionStorage.setItem('global_toolbar_mytool_ids', _self.clickCount.join(','));
          return false;
        }else{
          $(element).removeClass('checked');
          var ids = [];
          _self._getMyTools(function(err, data){
            var len = data.length;
            if(len > 6){
              len = len -1;
            }
            for(var i=0; i< len; i++){
              ids.push(data[i].id);
            }
            id = parseInt(id);
            if (ids.indexOf(id) === -1){
              ids.push(parseInt(id));
            }
            eventProxy.emit('tools.change', ids, _self._getSelTools(ids));
            eventProxy.emit('tools.destroy');
          });
        }
      });
    });
  },
  _getSelTools: function(ids){
    var _self = this;
    var plugins = [], i, len, id, item;
    for(i=0, len = ids.length; i< len; i++){
      id = parseInt(ids[i]);
      item = _self._getPlugin(id, _self.data.pluginList);
      plugins.push(item);
    }
    return plugins;
  },
  removeArr: function(arrays, index){
    var arr = [], len, left, right, i;
    len = arrays.length;
    if(len > 0){
      if(index === 0){
        return arrays.slice(1);
      }else if(index === len -1){
        return arrays.slice(0, len-1);
      }else{
        left = arrays.slice(0, index);
        right = arrays.slice(index + 1);
        for(i=0, len = left.length; i< len; i++){
          arr.push(left[i]);
        }
        for(i=0, len = right.length; i< len; i++){
          arr.push(right[i]);
        }
        return arr;
      }
    }
  },
  _pluginHover: function(){
    var _self = this,desc, userInfo, team;
    var workPrefix = 'http://work.alibaba-inc.com';
    this.checkboxArr.each(function(index, element){
      var id = $(element).attr('value');
      if(_self.data.pluginList){
        var obj = _self._getPlugin(parseInt(id), _self.data.pluginList);
        $(element).on('mouseover', function(){
          _self.pluginIcon.attr('class', '');
          _self.pluginIcon.addClass('pluginicon');
          _self.pluginIcon.addClass(obj.alias);
          _self.pluginIcon.parent().attr('class', '');
          _self.pluginIcon.parent().addClass(obj.alias+'psel');
          desc = $('.desc', _self.pluginDesc);
          team = $('.team', _self.pluginDesc);
          desc.html(obj.comment);
          if(obj.user_info){
            team.html('开发团队：' + obj.team);
            var objs = obj.user_info.split('|');
            var html = "<a target='_blank' href='"+workPrefix+ "/work/u/" +objs[0]+ "'>";
            html += "<img src='" +workPrefix+ "/photo/"+objs[0]+"'><label>"+objs[1]+"</label></a>";
            _self.photo.html(html);
          }else{
            team.html('');
            _self.photo.html('');
          }
        });
      }
    });
  },
  _getPlugin: function(id, pluginList){
    var _self = this;
    var obj = null, child;
    if(pluginList){
      for(var i=0, len = pluginList.length; i < len; i++){
        var item = pluginList[i];
        if(id === item.id){
          obj = item;
        }else{
          if(item.children){
            var children = item.children;
            for(var j=0, slen = children.length; j < slen; j++){
              child = children[j];
              if(id === child.id){
                obj = child;
              }
            }
          }
        }
      }
    }
    return obj;
  },
  _updateTools: function(ids, cb){
    var url = this.prefix + '/nr/header/updateMyTools';
    $.ajax({
      type: 'GET',
      url: url,
      data: {
        tools: ids
      },
      dataType:'jsonp',
      async: false,
      success: function(data){
        if(data){
          cb(null, JSON.parse(data));
        }
      },
      error: function(jqXHR, textStatus, errorThrown){
        cb(errorThrown);
      }
    });
  },
  _getTools: function(cb){
    var _self = this;
    var url = this.prefix + '/nr/header/tools';
    $.ajax({
      type: 'GET',
      url: url,
      dataType:'jsonp',
      async: false,
      success: function(data){
        localStorage.setItem('global_toolbar_all_tools', data);
        cb(null, _self.parseString(data));
      },
      error: function(jqXHR, textStatus, errorThrown){
        cb(errorThrown);
      }
    });
  },
  _getMyTools: function(cb){
    var url = this.prefix + '/nr/header/view';
    var _self = this;
    $.ajax({
      type: 'GET',
      url: url,
      dataType:'jsonp',
      async: false,
      success: function(data){
        if(data){
          data = data.replace(/\n|\r/g, '');
          data = data.replace(/\\/g, '\\\\');
          var obj = JSON.parse(data);
          cb(null, obj.myTools);
        }
      },
      error: function(jqXHR, textStatus, errorThrown){
        cb(errorThrown);
      }
    });
  },
  parseString: function(str){
    return (new Function( 'return (' + str + ')') ());
  },
  beautifyUrl: function(obj){
    var name = obj.name;
    var url = $(obj).attr('href') || '';
    var appId = localStorage.getItem('appId') || '';
    if(url.indexOf('{appId}') > -1){
      url = url.replace('{appId}', appId);
      if(appId === '' && name === 'ide'){
        url = url.replace(/\?.*/i, '');
      }
    }
    $(obj).attr('target', '_blank');
    $(obj).attr('href', url);
  }
});

exports = module.exports = Index;
