// webhook/run/sqlscan/res/temp/base.js
/**
 *  page/base 主页基类
 */
var Controller = require('hc-common/core/controller');
var $ = require('jquery');

/**
 *
 * class Base
 * extends core/controller
 *
 */
var Base = function(){};
Controller._extend(Base, {
  autoLayoutOnResize : false,
  /**
   * initComponent 初始化主页
   * @param  {object} config 应用配置数据
   *
   */
  initComponent : function(config){
    //维护一个组件列表
    this.coms = {};
    this.initEvents();
    this.readConfig(config);
    if(this.autoLayoutOnResize === true){
      this.initLayouter();
    }
  },
  initEvents : function(){
    this.events = {};
  },
  initLayouter : function(){
    var _self = this;
    $(window).on('resize', function(){
      _self.doResize();
    });
  },
  doResize : function(){
    var i, comp, size;
    size = {
      width : $(window).width(),
      height : $(window).height()
    };
    for(i in this.coms){
      comp = this.coms[i];
      if(comp.handler){
        comp.handler.doResize(size);
      }
    }
  },
  /**
   * hashChange 重写父类函数，hash值改变时查找响应组件
   * @param  {object} hash hash数据
   *
   */
  hashChange : function(hash){
    this.beforeLoadCom();
    if(hash.controller){
      this.loadCom(hash.controller);
    }
  },
  /**
   * beforeLoadCom 加载组件之前，先隐藏现有的组件。
   *
   */
  beforeLoadCom : function(){
    if(this.currentCom && this.currentCom.handler.holddisplay !== true){
      if(this.currentCom.handler.autoDestroy === true){
        this.currentCom.handler.destroy();
        this.currentCom.isLoad = false;
        this.currentCom.handler = null;
      }else{
        this.currentCom.handler.hide();
      }
    }
  },
  /**
   * readConfig 读取配置文件中的配置信息
   * @param  {object} config 组件配置数据
   *
   */
  readConfig : function(config){
    var coms = config.coms;
    var autoLoadComp = [], i, len;
    for(i in coms){
      this.coms[i] = this.createCom(i, coms[i]);
      if(autoLoadComp.length === 0 || coms[i].autoload === true){
        autoLoadComp.push(i);
      }
    }
    for(i = 0, len = autoLoadComp.length; i < len; i++){
      this.loadCom(autoLoadComp[i]);
    }
  },
  /**
   * createCom 生成组件
   * @param  {String} comId 组件名称
   * @param  {Object} config 组件配置项
   *
   */
  createCom : function(comId, config){
    var com = config.com;
    config.comId = comId;
    return {
      com : com,
      config : config,
      isLoad : false,
      handler : null
    };
  },
  /**
   * loadCom 加载组件
   * @param  {String} comId 组件名称
   *
   */
  loadCom : function(comId){
    var _self = this;
    var app = this.coms[comId];
    if(app){
      if(app.isLoad === false){
        return root.loadJs(app.com, function(err, objs){
          if(_self.events && _self.events[comId]){
            app.config.config.events = _self.events[comId];
          }
          app.config.config.router = _self.router.get();
          app.handler = objs[app.com]._create(app.config.config);
          app.handler.show();
          app.isLoad = true;
          _self.currentCom = app;
        });
      }
      app.handler.show();
      app.isLoad = true;
      this.currentCom = app;
    }
  }
});
exports = module.exports = Base;
