// webhook/run/sqlscan/res/page/common/fnmenu/fnmenu.js
var ep = require('event-pipe');
var $ = require('jquery');
var Tree = require('hc-common/tree/tree');
var Store = require('hc-common/core/store');
var context = require('../context');
var eventProxy = require('hc-common/core/eventproxy');
var FnMenu = function(){};
Tree._extend(FnMenu, {
  _className : 'fnmenu',
  // __filename : __filename,
  //菜单的参数，由viewport传入
  param : {},
  initEvents : function(){
    var _self = this;
    eventProxy.on('menu.switch', function(opt){
      _self.switchTree(opt);
    });
    eventProxy.on('hash.switch', function(opt){
      _self.emit('menuselected', opt);
    });
    this.on('selected', function(record, module, extendOpt){
      if(!extendOpt){
        extendOpt = {
          controller : record.controller
        };
      }
      _self.emit('menuselected', extendOpt);
    });
  },
  afterComplete : function(){
    FnMenu.prototype._super.afterComplete.call(this);
    this.dataTemplate = {
      '0': {
        'event' : false
      },
      'default': {
        'name' : 'name',
        'children' : 'children',
        'id' : 'controller',
        'event' : true
      }
    };
    this.update();
  },
  beforeSelected : function(el, record, module){
    return module.event;
  },
  update : function(){
    var _self = this;
    this.store.load({
      //url : context.urls.fnMenu,
      url : '/s/menu/',
      param : this.param,
      callback : function(){
        _self._activeItem();
      }
    });
  },
  _activeItem : function(){
    var item, opt;
    if(this.router && this.router.controller){
      opt = this.router;
      this.switchTree(opt);
      if(this.param.loadRullAdd && this.router.controller === this.param.loadRullAdd){
        this.emit('menuselected', opt);
      }
    }else{
      this.selectedByIndex(0);
    }
  }
});
exports = module.exports = FnMenu;
