// webhook/.cov/controllers/pack.js
/* ****** automatically generated by jscover - do not edit ******/
if (typeof _$jscoverage === 'undefined') { _$jscoverage = {}; }
/* ****** end - do not edit ******/
function BranchData() {
    this.position = -1;
    this.nodeLength = -1;
    this.src = null;
    this.evalFalse = 0;
    this.evalTrue = 0;

    this.init = function(position, nodeLength, src) {
        this.position = position;
        this.nodeLength = nodeLength;
        this.src = src;
        return this;
    }

    this.ranCondition = function(result) {
        if (result)
            this.evalTrue++;
        else
            this.evalFalse++;
    };

    this.pathsCovered = function() {
        var paths = 0;
        if (this.evalTrue > 0)
          paths++;
        if (this.evalFalse > 0)
          paths++;
        return paths;
    };

    this.covered = function() {
        return this.evalTrue > 0 && this.evalFalse > 0;
    };

    this.toJSON = function() {
        return '{"position":' + this.position
            + ',"nodeLength":' + this.nodeLength
            + ',"src":' + jscoverage_quote(this.src)
            + ',"evalFalse":' + this.evalFalse
            + ',"evalTrue":' + this.evalTrue + '}';
    };

    this.message = function() {
        if (this.evalTrue === 0 && this.evalFalse === 0)
            return 'Condition never evaluated         :\t' + this.src;
        else if (this.evalTrue === 0)
            return 'Condition never evaluated to true :\t' + this.src;
        else if (this.evalFalse === 0)
            return 'Condition never evaluated to false:\t' + this.src;
        else
            return 'Condition covered';
    };
}

BranchData.fromJson = function(jsonString) {
    var json = eval('(' + jsonString + ')');
    var branchData = new BranchData();
    branchData.init(json.position, json.nodeLength, json.src);
    branchData.evalFalse = json.evalFalse;
    branchData.evalTrue = json.evalTrue;
    return branchData;
};

BranchData.fromJsonObject = function(json) {
    var branchData = new BranchData();
    branchData.init(json.position, json.nodeLength, json.src);
    branchData.evalFalse = json.evalFalse;
    branchData.evalTrue = json.evalTrue;
    return branchData;
};

function buildBranchMessage(conditions) {
    var message = 'The following was not covered:';
    for (var i = 0; i < conditions.length; i++) {
        if (conditions[i] !== undefined && conditions[i] !== null && !conditions[i].covered())
          message += '\n- '+ conditions[i].message();
    }
    return message;
};

function convertBranchDataConditionArrayToJSON(branchDataConditionArray) {
    var array = [];
    var length = branchDataConditionArray.length;
    for (var condition = 0; condition < length; condition++) {
        var branchDataObject = branchDataConditionArray[condition];
        if (branchDataObject === undefined || branchDataObject === null) {
            value = 'null';
        } else {
            value = branchDataObject.toJSON();
        }
        array.push(value);
    }
    return '[' + array.join(',') + ']';
}

function convertBranchDataLinesToJSON(branchData) {
    if (branchData === undefined) {
        return '[]'
    }
    var array = [];
    var length = branchData.length;
    for (var line = 0; line < length; line++) {
        var branchDataObject = branchData[line];
        if (branchDataObject === undefined || branchDataObject === null) {
            value = 'null';
        } else {
            value = convertBranchDataConditionArrayToJSON(branchDataObject);
        }
        array.push(value);
    }
    return '[' + array.join(',') + ']';
}

function convertBranchDataLinesFromJSON(jsonObject) {
    if (jsonObject === undefined) {
        return [];
    }
    var length = jsonObject.length;
    for (var line = 0; line < length; line++) {
        var branchDataJSON = jsonObject[line];
        if (branchDataJSON !== null) {
            for (var conditionIndex = 0; conditionIndex < branchDataJSON.length; conditionIndex ++) {
                var condition = branchDataJSON[conditionIndex];
                if (condition !== null) {
                    branchDataJSON[conditionIndex] = BranchData.fromJsonObject(condition);
                }
            }
        }
    }
    return jsonObject;
}
try {
  if (typeof top === 'object' && top !== null && typeof top.opener === 'object' && top.opener !== null) {
    // this is a browser window that was opened from another window

    if (! top.opener._$jscoverage) {
      top.opener._$jscoverage = {};
      top.opener._$jscoverage.branchData = {};
    }
  }
}
catch (e) {}

try {
  if (typeof top === 'object' && top !== null) {
    // this is a browser window

    try {
      if (typeof top.opener === 'object' && top.opener !== null && top.opener._$jscoverage) {
        top._$jscoverage = top.opener._$jscoverage;
      }
    }
    catch (e) {}

    if (! top._$jscoverage) {
      top._$jscoverage = {};
      top._$jscoverage.branchData = {};
    }
  }
}
catch (e) {}

try {
  if (typeof top === 'object' && top !== null && top._$jscoverage) {
    this._$jscoverage = top._$jscoverage;
  }
}
catch (e) {}
if (! this._$jscoverage) {
  this._$jscoverage = {};
  this._$jscoverage.branchData = {};
}
if (! _$jscoverage['controllers/pack.js']) {
  _$jscoverage['controllers/pack.js'] = [];
  _$jscoverage['controllers/pack.js'][7] = 0;
  _$jscoverage['controllers/pack.js'][8] = 0;
  _$jscoverage['controllers/pack.js'][9] = 0;
  _$jscoverage['controllers/pack.js'][11] = 0;
  _$jscoverage['controllers/pack.js'][12] = 0;
  _$jscoverage['controllers/pack.js'][13] = 0;
  _$jscoverage['controllers/pack.js'][16] = 0;
  _$jscoverage['controllers/pack.js'][17] = 0;
  _$jscoverage['controllers/pack.js'][18] = 0;
  _$jscoverage['controllers/pack.js'][19] = 0;
  _$jscoverage['controllers/pack.js'][20] = 0;
  _$jscoverage['controllers/pack.js'][21] = 0;
  _$jscoverage['controllers/pack.js'][22] = 0;
  _$jscoverage['controllers/pack.js'][24] = 0;
  _$jscoverage['controllers/pack.js'][25] = 0;
  _$jscoverage['controllers/pack.js'][26] = 0;
  _$jscoverage['controllers/pack.js'][28] = 0;
  _$jscoverage['controllers/pack.js'][33] = 0;
  _$jscoverage['controllers/pack.js'][34] = 0;
}
_$jscoverage['controllers/pack.js'].source = ["/*!"," * /controllers/pack.js &#36127;&#36131;&#25171;&#21253;&#30340;&#20013;&#38388;&#20214;"," * Copyright(c) 2012"," * Author: jifeng &lt;wade428@163.com&gt;"," */","","var urlLib = require('url');","var parse = urlLib.parse;","var project = require('../lib/project');","","var Pack = function (options) {","  this.options = options;","  this.project = project(this.options)","}","","Pack.prototype.middleware = function() {","\tvar self = this;","\treturn function (req, res, next) {","    var urlObj = parse(req.url, true);","    var appname = urlObj.query.appname;","    if (!appname) {","      return res.end('appname is Invalid');","    }","    self.project.initRep(appname, function (err, info) {","      if (err) {","      \treturn next(err)","      }","      res.end('pack ' + appname + ' ok');","    });","\t};","};","","module.exports = function (options) {","\treturn new Pack(options);","}","",""];
_$jscoverage['controllers/pack.js'][7]++;
var urlLib = require('url');
_$jscoverage['controllers/pack.js'][8]++;
var parse = urlLib.parse;
_$jscoverage['controllers/pack.js'][9]++;
var project = require('../lib/project');
_$jscoverage['controllers/pack.js'][11]++;
var Pack = function(options) {
  _$jscoverage['controllers/pack.js'][12]++;
  this.options = options;
  _$jscoverage['controllers/pack.js'][13]++;
  this.project = project(this.options);
};
_$jscoverage['controllers/pack.js'][16]++;
Pack.prototype.middleware = function() {
  _$jscoverage['controllers/pack.js'][17]++;
  var self = this;
  _$jscoverage['controllers/pack.js'][18]++;
  return function(req, res, next) {
  _$jscoverage['controllers/pack.js'][19]++;
  var urlObj = parse(req.url, true);
  _$jscoverage['controllers/pack.js'][20]++;
  var appname = urlObj.query.appname;
  _$jscoverage['controllers/pack.js'][21]++;
  if (!appname) {
    _$jscoverage['controllers/pack.js'][22]++;
    return res.end('appname is Invalid');
  }
  _$jscoverage['controllers/pack.js'][24]++;
  self.project.initRep(appname, function(err, info) {
  _$jscoverage['controllers/pack.js'][25]++;
  if (err) {
    _$jscoverage['controllers/pack.js'][26]++;
    return next(err);
  }
  _$jscoverage['controllers/pack.js'][28]++;
  res.end('pack ' + appname + ' ok');
});
};
};
_$jscoverage['controllers/pack.js'][33]++;
module.exports = function(options) {
  _$jscoverage['controllers/pack.js'][34]++;
  return new Pack(options);
};
