// webhook/run/sqlscan/res/page/ruleview/ruleview.js
var Container = require('hc-common/container/container');
var context = require('../../temp/context');

var Ruleview = function(){};

Container._extend(Ruleview, {
  _className : 'ruleview',
  __filename: __filename,
  isHide : false,
  initData : function(){
    this.data = {
      url : context.urls
    };
  },
  initEvents : function(){
    this.searchEvent();
    this.rowClickEvent();
    this.tabEvent();
  },
  afterComplete : function(){
    Ruleview.prototype._super.afterComplete.call(this);
    var appId = this.query('combobox[name=appid]');
    appId.query = function(searchText){
      return {
        query : JSON.stringify({
          moduleId : 10,
          name : searchText
        })
      };
    };
    var owner = this.query('combobox[name=owner]');
    owner.query = function(searchText){
      return {
        query : JSON.stringify({
          keywords : $.trim(searchText)
        })
      };
    };
    var ruleName = this.query('combobox[name=rulename]');
    ruleName.query = function(searchText){
      return {
        query : searchText
      };
    };
    this.loadRulelist();
  },
  searchEvent : function(){
    var _self = this;
    var btnSearch = _self.query('button[name="btnSearch"]');
    btnSearch.on('click', function(){
      _self.loadRulelist();
    });
  },
  rowClickEvent : function(){
    var _self = this;
    var rulelist = _self.query('list[name="rulelist"]');
    rulelist.on('rowclick', function(record, row, e){
      if($(e.target).attr('class') === 'rule-desc'){
        _self.loadRuleDescDetail(record, $(e.target), this);
      }else if($(e.target).attr('value') === "authorization"){
        _self.viewAuthorizeRange(record);
      }
    })
  },
  tabEvent : function(){
    var _self = this;
    var tab = $('.rulelist-tab', this.el).children("span");
    tab.on('click' ,function(){
      if($(this).attr('class') === 'rulelist-tab-active'){
        return;
      }
      for(var i = 0, len = tab.length; i < len; i++){
        if(tab[i] === this){
          tab[i].className = "rulelist-tab-active"
        }else{
          tab[i].className = "rulelist-tab-deactive"
        }
      }
      _self.loadRulelist();
    });
  },
  loadRulelist : function(){
    var _self = this;
    var rulelist = _self.query('list[name="rulelist"]');
    var owner = _self.query('combobox[name="owner"]').val();
    var appId = _self.query('combobox[name="appid"]').val();
    var ruleName = _self.query('combobox[name="rulename"]').val();
    var status = _self.getActiveStatus();
    var opt = {
      url : context.urls.getRuleListView,
      param : {
        ruleName : ruleName,
        owner : owner,
        appId : appId,
        status : status
      },
      callback : function(){
        var count = rulelist.store.recordCount;
        var label = $('#rulecount');
        label.html('该条件下共有规则 ' + count +' 条');
        _self.emit('afterruleviewlistload');
      }
    };
    rulelist.store.loadPage(1, opt);
  },
  loadRuleDescDetail : function(record, desc, row){
    var detail = $(desc).parent().find(".rule-detail");
    if(record.detail.length <= 50){
      return;
    }
    if(!row.showDesc){
      //detail.html(record.activityDefinitionId);
      detail.slideDown();
      row.showDesc = true;
      this.checkRowClick = true;
      return;
    }
    detail.slideUp();
    row.showDesc = null;
    this.checkRowClick = null;
  },
  viewAuthorizeRange : function(record){
    var _self = this;
    root.loadJs('/page/authorizerange/authorizerange.js', function(err, objs){
      _self.win = objs['/page/authorizerange/authorizerange.js']._create({
        title : '授权范围',
        record : record
      });
      _self.emit('showauthorizerange');
    });
  },
  getActiveStatus : function(){
    var tab = $('.rulelist-tab');
    var status = tab.find('.rulelist-tab-active').attr('value');
    return status;
  }
});

exports = module.exports = Ruleview;
