// webhook/run/sqlscan/res/page/common/msgsetting/msgsetting.js
var WindowPanel = require('../../../com/window/windowpanel');
var context = require('../context');
var Store = require('../../../core/store');
var $ = require('jquery');

var MsgSetting = function(){};
WindowPanel._extend(MsgSetting, {
  _className : 'msgsetting',
  tplName: 'page/common/msgsetting/msgsetting',
  cssName: 'page/common/msgsetting/msgsetting',
  _loadStore : function(done){
    var _self = this;
    this.store = Store._create();
    this.store.load({
      'url' : context.urls.msgsetting,
      'param' : {},
      'callback': function(err, data){
        _self.defaultSetting = data;
        done();
      }
    });
  },
  afterComplete : function(){
    MsgSetting.prototype._super.afterComplete.call(this);
    this._container = $('.self-setting', this.body);
    this._settings = $('.item', this._container);
    this._loadDefaultSetting(this.defaultSetting);
  },
  _loadDefaultSetting : function(data){
    this._settings.each(function(i, item){
      var checks = $("input[type='checkbox']", item);
      var dt = data[$(item).attr('data')];
      checks.each(function(j, sub){
        if(dt.charAt(j) === "1"){
          sub.checked = true;
        }
      });
    });
  },
  _editSetting: function(){
    var data = {
      "workId": "TW000665"
    };
    this._settings.each(function(i, item){
      var checks = $("input[type='checkbox']", item);
      var values = [];
      checks.each(function(j, sub){
        if(sub.checked){
          values.push("1");
        }else{
          values.push("0");
        }
      });
      data[$(item).attr('data')] = values.join('');
    });
    var store = Store._create({param: JSON.stringify(data)});
    store.load({
      'url': context.urls.submit,
      'param': {},
      'callback': function(err, data){
        alert(data);
      }
    });
  },
  _editAccessInfo: function(){
    var _accessId = $('.access_id', this._container);
    var _accessKey = $('.access_key', this._container);
  },
  initEvents: function(){
    MsgSetting.prototype._super.initEvents.call(this);
    var btnCancel, btnSave, btnSaveAccess, _self;
    btnCancel = this.query("button[name='btnCancel']");
    btnSave = this.query("button[name='btnSave']");
    btnSaveAccess = this.query("button[name='btnSaveAccess']");
    _self = this;
    btnSaveAccess.on('click', function(){
      _self._editAccessInfo();
    });
    btnCancel.on('click', function(){
      _self.doClose();
    });
    btnSave.on('click', function(){
      _self._editSetting();
    });
  },
  destroy: function(){
    MsgSetting.prototype._super.destroy.call(this);
  }
});
exports = module.exports = MsgSetting;
