// webhook/.cov/run/sqlscan/res/page/ruleadd/authorizeitem/authorizeitem.js
var Container = require('hc-common/container/container');
var context = require('../../../temp/context');
var Msg = require('hc-common/msg/msg');
var Store = require('hc-common/core/store');

var AuthorizeItem = function(){};

Container._extend(AuthorizeItem, {
  _className : 'authorizeitem',
  __filename: __filename,
  isHide : false,
  initData : function(){
    this.authorizeitemIndex = $('.authorizecontainer').children().length;
    this.data = {
      url : context.urls,
      index : this.authorizeitemIndex
    };
    if(this.value){
      this.value = JSON.parse(this.value);
      this.data.proValue = this.value.projectName;
      this.data.usersValue = this.value.user;
    }
  },
  initEvents : function(){
    var _self = this;
    var btnRemove = this.query('button[name="btnRemove"]');
    btnRemove.on('click', function(){
      _self.destroy();
    });
  },
  initAppChangeEvents : function(app, owner){
    var _self = this;
    app.on('change', function(){
      var viewName = this._getRecord(this.val()).name;
      var cl, items = $('.authorizecontainer').find('input[placeholder=请选择应用名]');
      for(var i = 0, len = items.length; i < len; i++){
        cl = $(items[i]);
        if(cl[0] != this.el.find('input[placeholder=请选择应用名]')[0] && cl.val() === viewName){
          this.val('');
          Msg.alert('不能给重复应用');
          break;
        }
      }
      var appRecord = this._getRecord(this.val());
      _self.loadAppUsersStore(owner, appRecord);
    });
  },
  afterComplete : function(){
    AuthorizeItem.prototype._super.afterComplete.call(this);
    var _self = this, userInfo = context.options.session.user;
    var app = this.query('combobox[name=appId]');
    var owner = this.query('combobox[name=owner]');
    this.loadMyOwnerAppStore(app, userInfo.WorkId, function(){
      if(_self.value){
        app.val(_self.value.projectName);
        _self.loadAppUsersStore(owner, app._getRecord(app.val()), function(){
          if(_self.value){
            for(var i = 0, items = _self.value.user, len = items.length; i < len ; i++){
              owner.val(items[i].userName);
            }
            //之所以在compelte后再绑定事件，是需要在user加载完后再设置val
            _self.initAppChangeEvents(app, owner);
          }
        });
      }else{
        _self.loadAppUsersStore(owner);
        _self.initAppChangeEvents(app, owner);
      }
    });
  },
  myLoadStore : function(url, params, cb){
    var store = Store._create({
      url : url,
      param : params
    });
    store.load(function(err, data){
      if(err){
        return;
      }
      cb(data);
    });
  },
  loadMyOwnerAppStore : function(app, wokid, cb){
    var _self = this, params = {id : wokid, reauestId : '1'};
    this.myLoadStore(context.urls.myappInfos, params, function(rtn){
      var newData = [], data = rtn.records || rtn;
      for(var i = 0, len = data.length; i < len; i++){
        if(data[i].owner === wokid){
          newData.push(data[i]);
          continue;
        }
        for(var j = 0, roles = data[i].roles, count = roles.length; j < count; j++){
          if(roles[j].code === "PE"){
            newData.push(data[i]);
            break;
          }
        }
      }
      app.store.load(newData);
      cb();
      _self.emit('afterloadmyownerappstore');
    });
  },
  loadAppUsersStore : function(user, appRecord, cb){
    var _self = this, appId = appRecord ? appRecord.id : '';
    if(!appId){
      user.clear();
      user.store.loadLocalData([]);
      return;
    }
    var params = {
      query : JSON.stringify({
        appId : appId
      })
    };
    this.myLoadStore(context.urls.appUsers, params, function(data){
      user.clear();
      user.store.loadLocalData(data);
      if(cb){
        cb();
      }
      _self.emit('afterloadappusersstore');
    });
  },
  destroy : function(){
    AuthorizeItem.prototype._super.destroy.call(this);
    delete this.el;
  }
});

exports = module.exports = AuthorizeItem;
